<!DOCTYPE html>
<html lang="en"><head>
<script src="linear_regression_slides_files/libs/clipboard/clipboard.min.js"></script>
<script src="linear_regression_slides_files/libs/quarto-html/tabby.min.js"></script>
<script src="linear_regression_slides_files/libs/quarto-html/popper.min.js"></script>
<script src="linear_regression_slides_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="linear_regression_slides_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="linear_regression_slides_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="linear_regression_slides_files/libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.32">

  <meta name="author" content="Dr Stevie Pederson">
  <meta name="dcterms.date" content="2025-07-09">
  <title>Linear Regression in R</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="linear_regression_slides_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="linear_regression_slides_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    html { -webkit-text-size-adjust: 100%; }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="linear_regression_slides_files/libs/revealjs/dist/theme/quarto-30963254ca3fee1f61c50d8fd85c7cc4.css">
  <link href="linear_regression_slides_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="linear_regression_slides_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="linear_regression_slides_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="linear_regression_slides_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-color="#3d3d40" data-background-image="assets/bodl_logo_white_background.jpg" data-background-opacity="0.3" data-background-size="90%" class="quarto-title-block center">
  <h1 class="title">Linear Regression in R</h1>
  <p class="subtitle">RAdelaide 2025</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Dr Stevie Pederson 
</div>
        <p class="quarto-title-affiliation">
            </p><div class="line-block">Black Ochre Data Labs<br>
            Telethon Kids Institute</div>
          <p></p>
    </div>
</div>

  <p class="date">July 9, 2025</p>
</section>
<section>
<section id="linear-regression" class="title-slide slide level1 center" data-background-color="#3d3d40">
<h1>Linear Regression</h1>

</section>
<section id="setup" class="slide level2">
<h2>Setup</h2>
<ul>
<li>Start clear R session</li>
<li>Create a blank R script: <code>LinearRegression.R</code></li>
<li>Load the key packages</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(palmerpenguins)</span>
<span id="cb1-3"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(</span>
<span id="cb1-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb1-5"><a href="" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="linear-regression-1" class="slide level2">
<h2>Linear Regression</h2>
<p>We are trying to estimate a line with slope &amp; intercept</p>
<p><span class="math display">\[
y = ax + b
\]</span></p>
<div class="fragment">
<p>Or</p>
<p><span class="math display">\[
y = \beta_0 + \beta_1 x
\]</span></p>
<ul>
<li class="fragment"><span class="math inline">\(y\)</span> is the <em>response variable</em></li>
<li class="fragment"><span class="math inline">\(x\)</span> is the <em>predictor variable</em></li>
<li class="fragment">Makes the most intuitive sense when both <span class="math inline">\(x\)</span> &amp; <span class="math inline">\(y\)</span> are continuous</li>
</ul>
</div>
</section>
<section id="linear-regression-2" class="slide level2">
<h2>Linear Regression</h2>
<p>Linear Regression always uses the <code>R</code> formula syntax</p>
<ul>
<li><code>y ~ x</code>: <code>y</code> depends on <code>x</code></li>
<li>We use the function <code>lm()</code></li>
<li>Once we have our model, we can predict <span class="math inline">\(y\)</span> based on <span class="math inline">\(x\)</span> values</li>
</ul>
<div class="fragment">
<ul>
<li>We’ll use the penguins dataset for some exploration</li>
</ul>
</div>
</section>
<section id="linear-regression-3" class="slide level2">
<h2>Linear Regression</h2>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Does bill length depend on body mass?</span></span>
<span id="cb2-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Plot the predictor (body mass) on `x`</span></span>
<span id="cb2-3"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## The response (bill_length) goes on `y`</span></span>
<span id="cb2-4"><a href="" aria-hidden="true" tabindex="-1"></a>penguins <span class="sc">|&gt;</span></span>
<span id="cb2-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(</span>
<span id="cb2-6"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(body_mass_g, bill_length_mm)</span>
<span id="cb2-7"><a href="" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb2-8"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb2-9"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="linear_regression_slides_files/figure-revealjs/unnamed-chunk-2-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div></div>
<aside class="notes">
<ul>
<li>There looks like a fairly clear linear relationship</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="linear-regression-4" class="slide level2">
<h2>Linear Regression</h2>
<ul>
<li>Bill length is the response variable (<span class="math inline">\(y\)</span>)</li>
<li>Body Mass is the predictor variable (<span class="math inline">\(x\)</span>)</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Fit a linear regression model and save the output as an R object</span></span>
<span id="cb3-2"><a href="" aria-hidden="true" tabindex="-1"></a>bill_length_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(bill_length_mm <span class="sc">~</span> body_mass_g, <span class="at">data =</span> penguins)</span>
<span id="cb3-3"><a href="" aria-hidden="true" tabindex="-1"></a>bill_length_lm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = bill_length_mm ~ body_mass_g, data = penguins)

Coefficients:
(Intercept)  body_mass_g  
  26.898872     0.004051  </code></pre>
</div>
</div>
</section>
<section id="linear-regression-5" class="slide level2">
<h2>Linear Regression</h2>
<aside class="notes">
<ul>
<li>The basic output is only minimally informative</li>
<li>Often pass to the function <code>summary()</code> to present and interpret</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Use `summary` to view and interpret the fitted model</span></span>
<span id="cb5-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = bill_length_mm ~ body_mass_g, data = penguins)

Residuals:
     Min       1Q   Median       3Q      Max 
-10.1251  -3.0434  -0.8089   2.0711  16.1109 

Coefficients:
             Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept) 2.690e+01  1.269e+00   21.19   &lt;2e-16 ***
body_mass_g 4.051e-03  2.967e-04   13.65   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 4.394 on 340 degrees of freedom
  (2 observations deleted due to missingness)
Multiple R-squared:  0.3542,    Adjusted R-squared:  0.3523 
F-statistic: 186.4 on 1 and 340 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
</section>
<section id="linear-regression-6" class="slide level2">
<h2>Linear Regression</h2>
<ol type="1">
<li>The code used is returned as <code>Call:</code></li>
<li>A brief summary of the residuals are returned</li>
<li>The fitted values are returned in the <code>Coefficients</code> element
<ul>
<li>We have the estimate of the intercept &amp; slope</li>
<li>The standard error of the estimates</li>
<li>A t-test testing <span class="math inline">\(H_0: \beta_i = 0\)</span></li>
<li>The p-value for each t-test</li>
</ul></li>
<li>Additional model summary information
<ul>
<li><span class="math inline">\(R^2\)</span> is the proportion of variance explained by the model</li>
</ul></li>
</ol>
<aside class="notes">
<ul>
<li>Adjusted R-squared is the R-squared adjusted for the number of predictors</li>
<li>Mostly relevant when comparing models where predictors vary</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="coefficients" class="slide level2">
<h2>Coefficients</h2>
<div class="callout callout-important callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>How do we interpret the Intercept?</strong></p>
</div>
<div class="callout-content">
<div class="fragment">
<p>This is what the bill length would be if a penguin weighed exactly 0</p>
</div>
</div>
</div>
</div>
<div class="fragment">
<ul>
<li>We’d probably expect this to be zero but they almost never are</li>
<li>The Intercept is almost always significant</li>
<li>What does this really tell us about the relationship between bill length and weight?</li>
<li>Generally focussed on the relationship within the range of observed predictors
<ul>
<li>No guaranteed linear relationship outside of this range</li>
</ul></li>
</ul>
</div>
</section>
<section id="coefficients-1" class="slide level2">
<h2>Coefficients</h2>
<div class="callout callout-important callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>How do we interpret the <code>body_mass_g</code> term?</strong></p>
</div>
<div class="callout-content">
<div class="fragment">
<p>This is how much we would expect the bill length to change for <strong>every one unit increase in the predictor</strong> <br> i.e.&nbsp;for every 1<span class="math inline">\(g\)</span> increase in weight, the bill length would be expected to increase by about 0.0041mm</p>
</div>
</div>
</div>
</div>
<div class="fragment">
<ul>
<li>The <span class="math inline">\(t\)</span>-test here is highly relevant
<ul>
<li><span class="math inline">\(H_0\colon \beta_1 = 0~\)</span> Vs <span class="math inline">\(~H_A\colon \beta_1 \neq 0\)</span></li>
</ul></li>
<li>Reject <span class="math inline">\(H_0 \implies\)</span> there <em>is an association between predictor &amp; response</em></li>
</ul>
</div>
</section>
<section id="linear-regression-7" class="slide level2">
<h2>Linear Regression</h2>
<ul>
<li>Points never lie exactly on the regression line <span class="math inline">\(\implies\)</span> Why?</li>
</ul>
<div class="fragment">
<ul>
<li>We’re actually fitting the model</li>
</ul>
<p><span class="math display">\[
y_i = \beta_0 + \beta_1 x_i + \epsilon_i
\]</span></p>
<ul>
<li class="fragment"><span class="math inline">\(\beta_0 + \beta_1 x_i\)</span> is the exact line co-ordinate (Intercept + slope*predictor)</li>
<li class="fragment"><span class="math inline">\(\epsilon_i\)</span> is the the vertical difference between the observed value and the fitted value
<ul>
<li class="fragment">Known as a <em>residual</em></li>
<li class="fragment">Defined as <span class="math inline">\(\epsilon_i \sim \mathcal{N}(0, \sigma)\)</span></li>
</ul></li>
</ul>
<aside class="notes">
<ul>
<li>Residual means the bit left over when we subtract the predicted value from the observed</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="linear-regression-8" class="slide level2">
<h2>Linear Regression</h2>
<p>Linear Regression formally has 4 key assumptions</p>
<ol type="1">
<li class="fragment">Linear relationship between predictor and response
<ul>
<li class="fragment">Mean of residuals is zero across the entire range</li>
</ul></li>
<li class="fragment">Constant variance across the range of data (homoscedasticity)</li>
<li class="fragment">Residuals are normally distributed</li>
<li class="fragment">Independence of errors</li>
</ol>
<ul>
<li class="fragment">Three of these are represented in the definition <span class="math inline">\(\epsilon_i \sim \mathcal{N}(0, \sigma)\)</span></li>
</ul>
</section>
<section id="linear-regression-9" class="slide level2">
<h2>Linear Regression</h2>
<ul>
<li>To check our fitted model, we should check the residuals to ensure <span class="math inline">\(\epsilon_i \sim \mathcal{N}(0, \sigma)\)</span></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Check the residuals. It will ask you the following</span></span>
<span id="cb7-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Hit &lt;Return&gt; to see next plot: </span></span>
<span id="cb7-3"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## So follow the instructions and check each plot before proceeding</span></span>
<span id="cb7-4"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## There are 4 plots by default</span></span>
<span id="cb7-5"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(bill_length_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="linear-regression-10" class="slide level2">
<h2>Linear Regression</h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="linear_regression_slides_files/figure-revealjs/unnamed-chunk-6-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:40%;">
<div style="font-size: 90%">
<p><br></p>
<ul>
<li>Check the zero mean of <span class="math inline">\(\mathcal{N}(0, \sigma)\)</span></li>
<li>Is this assumption satisfied across the range of the data?</li>
</ul>
</div>
</div></div>
</section>
<section id="linear-regression-11" class="slide level2">
<h2>Linear Regression</h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="linear_regression_slides_files/figure-revealjs/unnamed-chunk-7-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:40%;">
<div style="font-size: 90%">
<p><br></p>
<ul>
<li>Check the normality of <span class="math inline">\(\mathcal{N}(0, \sigma)\)</span></li>
<li>The dashed line is the expected line from a normal distribution</li>
<li>Is this assumption satisfied across the range of the data?</li>
</ul>
</div>
</div></div>
</section>
<section id="linear-regression-12" class="slide level2">
<h2>Linear Regression</h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="linear_regression_slides_files/figure-revealjs/unnamed-chunk-8-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:40%;">
<div style="font-size: 90%">
<p><br></p>
<ul>
<li>Check the constant variance of <span class="math inline">\(\mathcal{N}(0, \sigma)\)</span></li>
<li>Is this assumption satisfied across the range of the data?</li>
</ul>
</div>
</div></div>
</section>
<section id="linear-regression-13" class="slide level2">
<h2>Linear Regression</h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="linear_regression_slides_files/figure-revealjs/unnamed-chunk-9-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:40%;">
<div style="font-size: 90%">
<p><br></p>
<ul>
<li>Checks if any points are exerting excessive ‘leverage’ on the model</li>
<li>Beyond scope of today</li>
</ul>
</div>
</div></div>
</section>
<section id="regression-diagnostic-plots" class="slide level2">
<h2>Regression Diagnostic Plots</h2>
<ul>
<li>All of these figures used base plotting functions</li>
<li>Stepping through can be frustrating
<ul>
<li>Especially when running an automated script</li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li>Can plot all 4 at once using a simple trick</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Define a 2x2 grid, plotting figures in a row-wise manner</span></span>
<span id="cb8-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb8-3"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Plot the four regression diagnostic plots</span></span>
<span id="cb8-4"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(bill_length_lm)</span>
<span id="cb8-5"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Reset the grid layout to be the default 1x1</span></span>
<span id="cb8-6"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>,))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<aside class="notes">
<ul>
<li><code>par()</code> is a function to set graphical parameters</li>
<li><code>mfrow</code> means ‘Multi-Figures By Row’</li>
<li>c(2, 2,) means plot in a 2x2 grid</li>
<li>I rarely show these so no need for ggplot versions</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="objects-of-class-lm" class="slide level2">
<h2>Objects Of Class <code>lm</code></h2>
<ul>
<li>The linear model we fitted produced an object of class <code>lm</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(bill_length_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "lm"</code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>This is also a list</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Inspect the actual object</span></span>
<span id="cb11-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(bill_length_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="objects-of-class-lm-1" class="slide level2">
<h2>Objects Of Class <code>lm</code></h2>
<ul>
<li>We can use the list structure to inspect the residuals manually</li>
</ul>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Plot a histogram of the residuals</span></span>
<span id="cb12-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(bill_length_lm<span class="sc">$</span>residuals)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="linear_regression_slides_files/figure-revealjs/unnamed-chunk-13-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div></div>
</section>
<section id="objects-of-class-lm-2" class="slide level2">
<h2>Objects Of Class <code>lm</code></h2>
<ul>
<li>We could even use the Shapiro-Wilk test for normality</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">shapiro.test</span>(bill_length_lm<span class="sc">$</span>residuals)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Shapiro-Wilk normality test

data:  bill_length_lm$residuals
W = 0.95439, p-value = 8.217e-09</code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>How can we interpret all of this?</li>
<li>Maybe there’s a better model</li>
</ul>
</div>
</section>
<section id="adding-terms" class="slide level2">
<h2>Adding Terms</h2>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Does bill length depend on body mass?</span></span>
<span id="cb15-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Plot the predictor (body mass) on `x`</span></span>
<span id="cb15-3"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## The response (bill_length) goes on `y`</span></span>
<span id="cb15-4"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Does each species need it's own model</span></span>
<span id="cb15-5"><a href="" aria-hidden="true" tabindex="-1"></a>penguins <span class="sc">|&gt;</span></span>
<span id="cb15-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(</span>
<span id="cb15-7"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb15-8"><a href="" aria-hidden="true" tabindex="-1"></a>      body_mass_g, bill_length_mm, </span>
<span id="cb15-9"><a href="" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour =</span> species</span>
<span id="cb15-10"><a href="" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb15-11"><a href="" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb15-12"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb15-13"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="linear_regression_slides_files/figure-revealjs/unnamed-chunk-15-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div></div>
<aside class="notes">
<ul>
<li>Do we think these slopes are the same?</li>
<li>Do we think the intercepts are the same?</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="adding-terms-1" class="slide level2">
<h2>Adding Terms</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Include the species in the model</span></span>
<span id="cb16-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## NB: This will fit a separate intercept for each species</span></span>
<span id="cb16-3"><a href="" aria-hidden="true" tabindex="-1"></a>bill_length_sp_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(bill_length_mm <span class="sc">~</span> species <span class="sc">+</span> body_mass_g, <span class="at">data =</span> penguins) </span>
<span id="cb16-4"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_sp_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = bill_length_mm ~ species + body_mass_g, data = penguins)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.8129 -1.6718  0.1336  1.4720  9.2902 

Coefficients:
                  Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)      2.492e+01  1.063e+00  23.443  &lt; 2e-16 ***
speciesChinstrap 9.921e+00  3.511e-01  28.258  &lt; 2e-16 ***
speciesGentoo    3.558e+00  4.858e-01   7.324 1.78e-12 ***
body_mass_g      3.749e-03  2.823e-04  13.276  &lt; 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 2.403 on 338 degrees of freedom
  (2 observations deleted due to missingness)
Multiple R-squared:  0.808, Adjusted R-squared:  0.8063 
F-statistic:   474 on 3 and 338 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<aside class="notes">
<ul>
<li>There is already a large increase in R-squared so this model may fit better</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="adding-terms-2" class="slide level2">
<h2>Adding Terms</h2>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Check the residuals after </span></span>
<span id="cb18-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## including a separate intercept</span></span>
<span id="cb18-3"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## for species</span></span>
<span id="cb18-4"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb18-5"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(bill_length_sp_lm)</span>
<span id="cb18-6"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="linear_regression_slides_files/figure-revealjs/unnamed-chunk-17-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div></div>
</section>
<section id="model-diagnostics" class="slide level2">
<h2>Model Diagnostics</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Check the histogram of residuals</span></span>
<span id="cb19-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(bill_length_sp_lm<span class="sc">$</span>residuals, <span class="at">breaks =</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Perform the Shapiro-Wilk test for Normality</span></span>
<span id="cb20-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">shapiro.test</span>(bill_length_sp_lm<span class="sc">$</span>residuals)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Shapiro-Wilk normality test

data:  bill_length_sp_lm$residuals
W = 0.99317, p-value = 0.123</code></pre>
</div>
</div>
</section>
<section id="interpreting-the-coefficients" class="slide level2">
<h2>Interpreting the Coefficients</h2>
<ul>
<li>Now we’re happier with the model <span class="math inline">\(\rightarrow\)</span> what do the coefficients mean?</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code>                     Estimate   Std. Error   t value     Pr(&gt;|t|)
(Intercept)      24.919470977 1.0630034684 23.442511 7.632983e-73
speciesChinstrap  9.920884113 0.3510790185 28.258265 5.093822e-91
speciesGentoo     3.557977539 0.4857896978  7.324111 1.776921e-12
body_mass_g       0.003748497 0.0002823439 13.276352 1.158990e-32</code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>By default, the primary Intercept is the first factor level in <code>species</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(penguins<span class="sc">$</span>species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Adelie"    "Chinstrap" "Gentoo"   </code></pre>
</div>
</div>
</div>
</section>
<section id="interpreting-the-coefficients-1" class="slide level2">
<h2>Interpreting the Coefficients</h2>
<ul>
<li>Now we’re happier with the model <span class="math inline">\(\rightarrow\)</span> what do the coefficients mean?</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code>                     Estimate   Std. Error   t value     Pr(&gt;|t|)
(Intercept)      24.919470977 1.0630034684 23.442511 7.632983e-73
speciesChinstrap  9.920884113 0.3510790185 28.258265 5.093822e-91
speciesGentoo     3.557977539 0.4857896978  7.324111 1.776921e-12
body_mass_g       0.003748497 0.0002823439 13.276352 1.158990e-32</code></pre>
</div>
</div>
<ul>
<li>The baseline intercept is for Adelie penguins</li>
<li>Additional intercept terms are the differences between the baseline and each species
<ul>
<li>Does this check-out in our initial plot of the data?</li>
</ul></li>
<li>They all appear significant when checking each <span class="math inline">\(H_0\)</span></li>
</ul>
</section>
<section id="adding-terms-3" class="slide level2">
<h2>Adding Terms</h2>
<ul>
<li>Do we think each species may have a different relationship between mass and bill length?
<ul>
<li>Do we need to fit a separate slope for each species?</li>
</ul></li>
<li>This is done in <code>R</code> using an “interaction term” separating terms by <code>:</code>
<ul>
<li><code>species:body_mass_g</code></li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="" aria-hidden="true" tabindex="-1"></a>bill_length_int_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(</span>
<span id="cb26-2"><a href="" aria-hidden="true" tabindex="-1"></a>  bill_length_mm <span class="sc">~</span> species <span class="sc">+</span> body_mass_g <span class="sc">+</span> species<span class="sc">:</span>body_mass_g, </span>
<span id="cb26-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> penguins</span>
<span id="cb26-4"><a href="" aria-hidden="true" tabindex="-1"></a>) </span>
<span id="cb26-5"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_int_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = bill_length_mm ~ species + body_mass_g + species:body_mass_g, 
    data = penguins)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.4208 -1.6461  0.0919  1.4718  9.3138 

Coefficients:
                               Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)                  26.9941391  1.5926015  16.950  &lt; 2e-16 ***
speciesChinstrap              5.1800537  3.2746719   1.582    0.115    
speciesGentoo                -0.2545907  2.7138655  -0.094    0.925    
body_mass_g                   0.0031879  0.0004271   7.464 7.27e-13 ***
speciesChinstrap:body_mass_g  0.0012748  0.0008740   1.459    0.146    
speciesGentoo:body_mass_g     0.0009030  0.0006066   1.489    0.138    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 2.399 on 336 degrees of freedom
  (2 observations deleted due to missingness)
Multiple R-squared:  0.8098,    Adjusted R-squared:  0.807 
F-statistic: 286.1 on 5 and 336 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<aside class="notes">
<ul>
<li>Now what do we think?</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="interpreting-the-coefficients-2" class="slide level2">
<h2>Interpreting the Coefficients</h2>
<ul>
<li>The baseline terms for the Intercept and <code>body_mass_g</code> are now <strong>both</strong> for Adelie</li>
<li>Differences in slope for each species are provided as <code>speciesChinstrap:body_mass_g</code> and <code>speciesGentoo:body_mass_g</code></li>
</ul>
<div class="fragment">
<ul>
<li>The regression line for Adelie is y = 26.99 + 0.0032*<code>body_mass_g</code></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>For Chinstrap: y = (26.99+5.18) + (0.0032+0.0013)*body_mass_g`</li>
</ul>
</div>
</section>
<section id="interpreting-the-coefficients-3" class="slide level2">
<h2>Interpreting the Coefficients</h2>
<ul>
<li>The model matrix <span class="math inline">\(X\)</span> is a key part of understanding statistics</li>
<li>The model coefficients (<span class="math inline">\(\hat{\beta}\)</span>) are actually estimated by performing linear algebra on this</li>
<li>The least squares solution is <span class="math inline">\(\hat{\beta} = (X^TX)^{-1}X^Ty\)</span></li>
</ul>
<div class="fragment">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> penguins<span class="sc">$</span>bill_length_mm[<span class="sc">!</span><span class="fu">is.na</span>(penguins<span class="sc">$</span>body_mass_g)]</span>
<span id="cb28-2"><a href="" aria-hidden="true" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> species <span class="sc">+</span> body_mass_g <span class="sc">+</span> species<span class="sc">:</span>body_mass_g, <span class="at">data =</span> penguins)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="interpreting-the-coefficients-4" class="slide level2">
<h2>Interpreting the Coefficients</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  (Intercept) speciesChinstrap speciesGentoo body_mass_g
1           1                0             0        3750
2           1                0             0        3800
3           1                0             0        3250
5           1                0             0        3450
6           1                0             0        3650
7           1                0             0        3625
  speciesChinstrap:body_mass_g speciesGentoo:body_mass_g
1                            0                         0
2                            0                         0
3                            0                         0
5                            0                         0
6                            0                         0
7                            0                         0</code></pre>
</div>
</div>
</section>
<section id="interpreting-the-coefficients-5" class="slide level2">
<h2>Interpreting the Coefficients</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## No need to save this, I'm just demonstrating a point</span></span>
<span id="cb31-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## The inverse of a matrix is found using `solve()`</span></span>
<span id="cb31-3"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Matrix multiplication is performed using %*%</span></span>
<span id="cb31-4"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## The transpose of a matrix is found using `t()`</span></span>
<span id="cb31-5"><a href="" aria-hidden="true" tabindex="-1"></a>beta_hat <span class="ot">&lt;-</span> <span class="fu">solve</span>(<span class="fu">t</span>(X) <span class="sc">%*%</span> X) <span class="sc">%*%</span> <span class="fu">t</span>(X) <span class="sc">%*%</span> y</span>
<span id="cb31-6"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">cbind</span>(beta_hat, <span class="fu">coef</span>(bill_length_int_lm))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                      [,1]          [,2]
(Intercept)                  26.9941391366 26.9941391367
speciesChinstrap              5.1800537287  5.1800537287
speciesGentoo                -0.2545906615 -0.2545906615
body_mass_g                   0.0031878758  0.0031878758
speciesChinstrap:body_mass_g  0.0012748183  0.0012748183
speciesGentoo:body_mass_g     0.0009029956  0.0009029956</code></pre>
</div>
</div>
</section>
<section id="interpreting-the-coefficients-6" class="slide level2">
<h2>Interpreting the Coefficients</h2>
<div style="font-size: 85%">
<ul>
<li>An excellent primer on model matrices is <a href="https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html">here</a><sup>1</sup></li>
<li>Conventional statistics always sets a baseline level for the intercept
<ul>
<li>Performs a <span class="math inline">\(t\)</span>-test automatically for differences</li>
<li>Really quite convenient if less obvious</li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li>We could remove the common intercept using <code>~ 0 + species + body_mass_g.</code>
<ul>
<li>Would return a separate intercept by removing the common (baseline) level</li>
<li>No real advantage except non-statisticians like it</li>
</ul></li>
<li>No statistical evidence provided for differences in intercepts<br><span class="math inline">\(\implies\)</span> have to test ourselves 😢</li>
</ul>
</div>
</div>
<aside><ol class="aside-footnotes"><li id="fn1"><p>https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html</p></li></ol></aside></section>
<section id="model-diagnostics-1" class="slide level2">
<h2>Model Diagnostics</h2>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Check the residuals after </span></span>
<span id="cb33-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## including a separate intercept</span></span>
<span id="cb33-3"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## and slope for species</span></span>
<span id="cb33-4"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb33-5"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(bill_length_int_lm)</span>
<span id="cb33-6"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="linear_regression_slides_files/figure-revealjs/unnamed-chunk-26-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div></div>
</section>
<section id="model-selection" class="slide level2">
<h2>Model Selection</h2>
<ul>
<li>How do we decide on the best model?</li>
<li>A common technique is Analysis of Variance (ANOVA)</li>
<li>Classic ANOVA checks importance of each term within a model</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Run a separate ANOVA on each model</span></span>
<span id="cb34-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(bill_length_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: bill_length_mm
             Df Sum Sq Mean Sq F value    Pr(&gt;F)    
body_mass_g   1 3599.7  3599.7  186.44 &lt; 2.2e-16 ***
Residuals   340 6564.5    19.3                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(bill_length_sp_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: bill_length_mm
             Df Sum Sq Mean Sq F value    Pr(&gt;F)    
species       2 7194.3  3597.2  622.88 &lt; 2.2e-16 ***
body_mass_g   1 1017.9  1017.9  176.26 &lt; 2.2e-16 ***
Residuals   338 1952.0     5.8                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(bill_length_int_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: bill_length_mm
                     Df Sum Sq Mean Sq  F value Pr(&gt;F)    
species               2 7194.3  3597.2 625.1488 &lt;2e-16 ***
body_mass_g           1 1017.9  1017.9 176.9039 &lt;2e-16 ***
species:body_mass_g   2   18.6     9.3   1.6159 0.2003    
Residuals           336 1933.4     5.8                    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>Does this give any clue as to the best model?</li>
</ul>
</div>
</section>
<section id="model-selection-1" class="slide level2">
<h2>Model Selection</h2>
<ul>
<li>We have progressively added terms to each model</li>
<li>Can use ANOVA to compare suitability of each model</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Compare the models using ANOVA</span></span>
<span id="cb40-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## If a model is a significant improvement, the p-value from the </span></span>
<span id="cb40-3"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## F test will be clearly significant</span></span>
<span id="cb40-4"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(</span>
<span id="cb40-5"><a href="" aria-hidden="true" tabindex="-1"></a>  bill_length_lm,</span>
<span id="cb40-6"><a href="" aria-hidden="true" tabindex="-1"></a>  bill_length_sp_lm,</span>
<span id="cb40-7"><a href="" aria-hidden="true" tabindex="-1"></a>  bill_length_int_lm</span>
<span id="cb40-8"><a href="" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Model 1: bill_length_mm ~ body_mass_g
Model 2: bill_length_mm ~ species + body_mass_g
Model 3: bill_length_mm ~ species + body_mass_g + species:body_mass_g
  Res.Df    RSS Df Sum of Sq        F Pr(&gt;F)    
1    340 6564.5                                 
2    338 1952.0  2    4612.5 400.8045 &lt;2e-16 ***
3    336 1933.4  2      18.6   1.6159 0.2003    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>It looks like the separate slopes are not an improvement</li>
<li>The separate intercepts <strong>are</strong> an improvement</li>
</ul>
</div>
</section>
<section id="speeding-the-process-up" class="slide level2">
<h2>Speeding The Process Up</h2>
<ul>
<li>This was a careful breakdown of finding the best model</li>
<li>We can partially automate this and use some shortcuts</li>
</ul>
<div class="fragment">
<ul>
<li>The shorthand for an interaction term with separate intercepts is <code>*</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Refit the model with an interaction term using the shorthand code</span></span>
<span id="cb42-2"><a href="" aria-hidden="true" tabindex="-1"></a>bill_length_int_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(</span>
<span id="cb42-3"><a href="" aria-hidden="true" tabindex="-1"></a>  bill_length_mm <span class="sc">~</span> species <span class="sc">*</span> body_mass_g, <span class="at">data =</span> penguins</span>
<span id="cb42-4"><a href="" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb42-5"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_int_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = bill_length_mm ~ species * body_mass_g, data = penguins)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.4208 -1.6461  0.0919  1.4718  9.3138 

Coefficients:
                               Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)                  26.9941391  1.5926015  16.950  &lt; 2e-16 ***
speciesChinstrap              5.1800537  3.2746719   1.582    0.115    
speciesGentoo                -0.2545907  2.7138655  -0.094    0.925    
body_mass_g                   0.0031879  0.0004271   7.464 7.27e-13 ***
speciesChinstrap:body_mass_g  0.0012748  0.0008740   1.459    0.146    
speciesGentoo:body_mass_g     0.0009030  0.0006066   1.489    0.138    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 2.399 on 336 degrees of freedom
  (2 observations deleted due to missingness)
Multiple R-squared:  0.8098,    Adjusted R-squared:  0.807 
F-statistic: 286.1 on 5 and 336 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
</div>
<div class="fragment">
<ul>
<li>Alternatively, all terms can be placed inside brackets &amp; raised to a power
<ul>
<li><code>(species + body_mass_g)^2</code> would give two-way interactions</li>
</ul></li>
</ul>
</div>
</section>
<section id="speeding-the-process-up-1" class="slide level2">
<h2>Speeding The Process Up</h2>
<aside class="notes">
<ul>
<li>The AIC is a function of the number of model terms and the log likelihood function</li>
<li>Beyond the scope of today</li>
<li>Doesn’t perform well with highly correlated predictors</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<ul>
<li>After specifying a ‘full’ model <span class="math inline">\(\implies\)</span> use <code>step()</code> to remove redundant terms
<ul>
<li>Removes terms in a step-wise manner</li>
<li>Uses Akaike’s Information Criterion (AIC) to determine optimal model</li>
<li>Finds model with lowest AIC</li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">step</span>(bill_length_int_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Start:  AIC=604.42
bill_length_mm ~ species * body_mass_g

                      Df Sum of Sq    RSS    AIC
- species:body_mass_g  2    18.596 1952.0 603.69
&lt;none&gt;                             1933.4 604.42

Step:  AIC=603.69
bill_length_mm ~ species + body_mass_g

              Df Sum of Sq    RSS     AIC
&lt;none&gt;                     1952.0  603.69
- body_mass_g  1    1017.9 2969.9  745.22
- species      2    4612.5 6564.5 1014.48</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = bill_length_mm ~ species + body_mass_g, data = penguins)

Coefficients:
     (Intercept)  speciesChinstrap     speciesGentoo       body_mass_g  
       24.919471          9.920884          3.557978          0.003748  </code></pre>
</div>
</div>
</section>
<section id="objects-of-class-summary.lm" class="slide level2">
<h2>Objects of Class <code>summary.lm</code></h2>
<ul>
<li>The coefficients element of the basic <code>lm</code> object only had the fitted values
<ul>
<li>Not the std errors, t-tests or p-values</li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Extract the coefficients directly from the linear model</span></span>
<span id="cb47-2"><a href="" aria-hidden="true" tabindex="-1"></a>bill_length_sp_lm<span class="sc">$</span>coefficients</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     (Intercept) speciesChinstrap    speciesGentoo      body_mass_g 
    24.919470977      9.920884113      3.557977539      0.003748497 </code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>These values are produced by the function <code>summary()</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Extract the coefficients from the summary object</span></span>
<span id="cb49-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_sp_lm)<span class="sc">$</span>coefficients</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                     Estimate   Std. Error   t value     Pr(&gt;|t|)
(Intercept)      24.919470977 1.0630034684 23.442511 7.632983e-73
speciesChinstrap  9.920884113 0.3510790185 28.258265 5.093822e-91
speciesGentoo     3.557977539 0.4857896978  7.324111 1.776921e-12
body_mass_g       0.003748497 0.0002823439 13.276352 1.158990e-32</code></pre>
</div>
</div>
</div>
</section>
<section id="objects-of-class-summary.lm-1" class="slide level2">
<h2>Objects of Class <code>summary.lm</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Check the class of the output from summary</span></span>
<span id="cb51-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_sp_lm) <span class="sc">|&gt;</span> <span class="fu">class</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "summary.lm"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Prove conclusively that it is really a list</span></span>
<span id="cb53-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## with a class attribute stuck on it</span></span>
<span id="cb53-3"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_sp_lm) <span class="sc">|&gt;</span> <span class="fu">is.list</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## See what attributes it has</span></span>
<span id="cb55-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_sp_lm) <span class="sc">|&gt;</span> <span class="fu">attributes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$names
 [1] "call"          "terms"         "residuals"     "coefficients" 
 [5] "aliased"       "sigma"         "df"            "r.squared"    
 [9] "adj.r.squared" "fstatistic"    "cov.unscaled"  "na.action"    

$class
[1] "summary.lm"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Check the full details</span></span>
<span id="cb57-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_sp_lm) <span class="sc">|&gt;</span> <span class="fu">glimpse</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 12
 $ call         : language lm(formula = bill_length_mm ~ species + body_mass_g, data = penguins)
 $ terms        :Classes 'terms', 'formula'  language bill_length_mm ~ species + body_mass_g
  .. ..- attr(*, "variables")= language list(bill_length_mm, species, body_mass_g)
  .. ..- attr(*, "factors")= int [1:3, 1:2] 0 1 0 0 0 1
  .. .. ..- attr(*, "dimnames")=List of 2
  .. ..- attr(*, "term.labels")= chr [1:2] "species" "body_mass_g"
  .. ..- attr(*, "order")= int [1:2] 1 1
  .. ..- attr(*, "intercept")= int 1
  .. ..- attr(*, "response")= int 1
  .. ..- attr(*, ".Environment")=&lt;environment: R_GlobalEnv&gt; 
  .. ..- attr(*, "predvars")= language list(bill_length_mm, species, body_mass_g)
  .. ..- attr(*, "dataClasses")= Named chr [1:3] "numeric" "factor" "numeric"
  .. .. ..- attr(*, "names")= chr [1:3] "bill_length_mm" "species" "body_mass_g"
 $ residuals    : Named num [1:342] 0.124 0.336 3.198 -1.152 0.699 ...
  ..- attr(*, "names")= chr [1:342] "1" "2" "3" "5" ...
 $ coefficients : num [1:4, 1:4] 24.91947 9.92088 3.55798 0.00375 1.063 ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : chr [1:4] "(Intercept)" "speciesChinstrap" "speciesGentoo" "body_mass_g"
  .. ..$ : chr [1:4] "Estimate" "Std. Error" "t value" "Pr(&gt;|t|)"
 $ aliased      : Named logi [1:4] FALSE FALSE FALSE FALSE
  ..- attr(*, "names")= chr [1:4] "(Intercept)" "speciesChinstrap" "speciesGentoo" "body_mass_g"
 $ sigma        : num 2.4
 $ df           : int [1:3] 4 338 4
 $ r.squared    : num 0.808
 $ adj.r.squared: num 0.806
 $ fstatistic   : Named num [1:3] 474 3 338
  ..- attr(*, "names")= chr [1:3] "value" "numdf" "dendf"
 $ cov.unscaled : num [1:4, 1:4] 1.96e-01 -4.97e-03 6.36e-02 -5.11e-05 -4.97e-03 ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : chr [1:4] "(Intercept)" "speciesChinstrap" "speciesGentoo" "body_mass_g"
  .. ..$ : chr [1:4] "(Intercept)" "speciesChinstrap" "speciesGentoo" "body_mass_g"
 $ na.action    : 'omit' Named int [1:2] 4 272
  ..- attr(*, "names")= chr [1:2] "4" "272"
 - attr(*, "class")= chr "summary.lm"</code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>The full complexity of the object is mostly irrelevant</li>
</ul>
</div>
</section>
<section id="using-a-more-tidyverse-friendly-approach" class="slide level2">
<h2>Using A More Tidyverse Friendly Approach</h2>
<ul>
<li>The function <code>tidy()</code> from the package <code>broom</code> is a catch-all function
<ul>
<li>Will return a tibble</li>
<li>Returns the same from <code>lm</code> and <code>summary.lm</code> objects</li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="" aria-hidden="true" tabindex="-1"></a>bill_length_sp_lm <span class="sc">|&gt;</span> broom<span class="sc">::</span><span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  term             estimate std.error statistic  p.value
  &lt;chr&gt;               &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)      24.9      1.06         23.4  7.63e-73
2 speciesChinstrap  9.92     0.351        28.3  5.09e-91
3 speciesGentoo     3.56     0.486         7.32 1.78e-12
4 body_mass_g       0.00375  0.000282     13.3  1.16e-32</code></pre>
</div>
</div>
</section>
<section id="adding-significance-stars" class="slide level2">
<h2>Adding Significance Stars</h2>
<ul>
<li>The easiest way for me as a <code>case_when()</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="" aria-hidden="true" tabindex="-1"></a>bill_length_sp_lm <span class="sc">|&gt;</span> </span>
<span id="cb61-2"><a href="" aria-hidden="true" tabindex="-1"></a>  broom<span class="sc">::</span><span class="fu">tidy</span>() <span class="sc">|&gt;</span></span>
<span id="cb61-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb61-4"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">stars =</span> <span class="fu">case_when</span>(</span>
<span id="cb61-5"><a href="" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.001</span> <span class="sc">~</span> <span class="st">"***"</span>,</span>
<span id="cb61-6"><a href="" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.01</span> <span class="sc">~</span> <span class="st">"**"</span>,</span>
<span id="cb61-7"><a href="" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">~</span> <span class="st">"*"</span>,</span>
<span id="cb61-8"><a href="" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.1</span> <span class="sc">~</span> <span class="st">"."</span>,</span>
<span id="cb61-9"><a href="" aria-hidden="true" tabindex="-1"></a>      <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">""</span></span>
<span id="cb61-10"><a href="" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb61-11"><a href="" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 6
  term             estimate std.error statistic  p.value stars
  &lt;chr&gt;               &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;
1 (Intercept)      24.9      1.06         23.4  7.63e-73 ***  
2 speciesChinstrap  9.92     0.351        28.3  5.09e-91 ***  
3 speciesGentoo     3.56     0.486         7.32 1.78e-12 ***  
4 body_mass_g       0.00375  0.000282     13.3  1.16e-32 ***  </code></pre>
</div>
</div>
</section>
<section id="or-fitting-a-model-on-the-fly" class="slide level2">
<h2>Or Fitting a Model On The Fly</h2>
<ul>
<li>Obviously no room for checking model diagnostics</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="" aria-hidden="true" tabindex="-1"></a>penguins <span class="sc">|&gt;</span></span>
<span id="cb63-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Piped data can be recalled using `_`</span></span>
<span id="cb63-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lm</span>( bill_length_mm <span class="sc">~</span> species <span class="sc">*</span> body_mass_g, <span class="at">data =</span> _) <span class="sc">|&gt;</span> </span>
<span id="cb63-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">step</span>() <span class="sc">|&gt;</span> <span class="co"># Fit the best model using the AIC</span></span>
<span id="cb63-5"><a href="" aria-hidden="true" tabindex="-1"></a>  broom<span class="sc">::</span><span class="fu">tidy</span>() <span class="sc">|&gt;</span> <span class="co"># Turn the output into a tibble &amp; add stars</span></span>
<span id="cb63-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb63-7"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">stars =</span> <span class="fu">case_when</span>(</span>
<span id="cb63-8"><a href="" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.001</span> <span class="sc">~</span> <span class="st">"***"</span>,</span>
<span id="cb63-9"><a href="" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.01</span> <span class="sc">~</span> <span class="st">"**"</span>,</span>
<span id="cb63-10"><a href="" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">~</span> <span class="st">"*"</span>,</span>
<span id="cb63-11"><a href="" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.1</span> <span class="sc">~</span> <span class="st">"."</span>,</span>
<span id="cb63-12"><a href="" aria-hidden="true" tabindex="-1"></a>      <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">""</span></span>
<span id="cb63-13"><a href="" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb63-14"><a href="" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Start:  AIC=604.42
bill_length_mm ~ species * body_mass_g

                      Df Sum of Sq    RSS    AIC
- species:body_mass_g  2    18.596 1952.0 603.69
&lt;none&gt;                             1933.4 604.42

Step:  AIC=603.69
bill_length_mm ~ species + body_mass_g

              Df Sum of Sq    RSS     AIC
&lt;none&gt;                     1952.0  603.69
- body_mass_g  1    1017.9 2969.9  745.22
- species      2    4612.5 6564.5 1014.48</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 6
  term             estimate std.error statistic  p.value stars
  &lt;chr&gt;               &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;
1 (Intercept)      24.9      1.06         23.4  7.63e-73 ***  
2 speciesChinstrap  9.92     0.351        28.3  5.09e-91 ***  
3 speciesGentoo     3.56     0.486         7.32 1.78e-12 ***  
4 body_mass_g       0.00375  0.000282     13.3  1.16e-32 ***  </code></pre>
</div>
</div>
</section></section>
<section>
<section id="s3-method-dispatch" class="title-slide slide level1 center">
<h1>S3 Method Dispatch</h1>

</section>
<section id="s3-method-dispatch-1" class="slide level2">
<h2>S3 Method Dispatch</h2>
<ul>
<li>Objects with class <code>lm</code> and <code>summary.lm</code> are <code>S3</code> objects</li>
<li>Very informal class structure in <code>R</code></li>
<li>Easy to work with <span class="math inline">\(\implies\)</span> easy to break</li>
</ul>
<div class="fragment">
<ul>
<li>When we printed these objects <span class="math inline">\(\implies\)</span> <code>print.lm()</code> or <code>print.summary.lm()</code></li>
<li>Likewise when we plotted the <code>lm</code> object <span class="math inline">\(\implies\)</span> <code>plot.lm()</code></li>
</ul>
</div>
<div class="fragment">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="" aria-hidden="true" tabindex="-1"></a>?plot.lm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="s3-method-dispatch-2" class="slide level2">
<h2>S3 Method Dispatch</h2>
<ul>
<li>If we only want a subset of figures</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb67-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(</span>
<span id="cb67-3"><a href="" aria-hidden="true" tabindex="-1"></a>  bill_length_sp_lm, <span class="at">which =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>),</span>
<span id="cb67-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">caption =</span> <span class="fu">list</span>(</span>
<span id="cb67-5"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Separate Intercepts: Residuals vs Fitted"</span>, <span class="cn">NULL</span>,</span>
<span id="cb67-6"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Separate Intercepts: Scale-Location"</span></span>
<span id="cb67-7"><a href="" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb67-8"><a href="" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<img data-src="linear_regression_slides_files/figure-revealjs/unnamed-chunk-38-1.png" class="quarto-figure quarto-figure-center r-stretch" width="768"></section>
<section id="s3-method-dispatch-3" class="slide level2">
<h2>S3 Method Dispatch</h2>
<ul>
<li>We can also check what we pass to <code>broom::tidy()</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="" aria-hidden="true" tabindex="-1"></a>?broom<span class="sc">::</span>tidy.lm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<ul>
<li><code>broom::tidy()</code> has multiple versions for objects of different classes
<ul>
<li>e.g.&nbsp;<code>broom::tidy.prcomp()</code> for PCA results</li>
</ul></li>
</ul>
</div>
</section>
<section id="s3-method-dispatch-4" class="slide level2">
<h2>S3 Method Dispatch</h2>
<ul>
<li><code>S3</code> objects can be easily broken</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="" aria-hidden="true" tabindex="-1"></a>broken_lm <span class="ot">&lt;-</span> bill_length_sp_lm</span>
<span id="cb70-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(broken_lm) <span class="ot">&lt;-</span> <span class="st">"htest"</span></span>
<span id="cb70-3"><a href="" aria-hidden="true" tabindex="-1"></a>broken_lm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>


data:  </code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li><code>R</code> looked for the <code>print.htest</code> method</li>
<li>The structure didn’t match what was expected <span class="math inline">\(\implies\)</span> nonsense output</li>
</ul>
</div>
</section>
<section id="confidence-intervals" class="slide level2">
<h2>Confidence Intervals</h2>
<ul>
<li>For confidence or prediction intervals, we can use <code>predict.lm()</code></li>
</ul>
<div class="callout callout-important callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>What’s the difference between a confidence and prediction interval</strong></p>
</div>
<div class="callout-content">
<div class="fragment">
<ul>
<li>95% Confidence Intervals: The mean will be in the given interval 95% of the time</li>
<li>95% Prediction Intervals: An observation will be in the given interval 95% of the time</li>
</ul>
</div>
</div>
</div>
</div>
<div class="fragment">
<ul>
<li>We will need a new data frame to make the predictions about</li>
</ul>
</div>
</section>
<section id="confidence-intervals-1" class="slide level2">
<h2>Confidence Intervals</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Create some new penguins</span></span>
<span id="cb72-2"><a href="" aria-hidden="true" tabindex="-1"></a>new_penguins <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb72-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">species =</span> <span class="fu">c</span>(<span class="st">"Adelie"</span>, <span class="st">"Gentoo"</span>), </span>
<span id="cb72-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">body_mass_g =</span> <span class="dv">4500</span></span>
<span id="cb72-5"><a href="" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb72-6"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Predict their mean bill length</span></span>
<span id="cb72-7"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">predict</span>(bill_length_sp_lm, <span class="at">newdata =</span> new_penguins, <span class="at">interval =</span> <span class="st">"confidence"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       fit      lwr      upr
1 41.78771 41.20030 42.37512
2 45.34568 44.81277 45.87860</code></pre>
</div>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Predict the range a new observation may lie in</span></span>
<span id="cb74-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">predict</span>(bill_length_sp_lm, <span class="at">newdata =</span> new_penguins, <span class="at">interval =</span> <span class="st">"prediction"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       fit      lwr      upr
1 41.78771 37.02436 46.55105
2 45.34568 40.58876 50.10261</code></pre>
</div>
</div>
</section>
<section id="confidence-intervals-2" class="slide level2">
<h2>Confidence Intervals</h2>
<ul>
<li>We may wish to include our new penguins in these results</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Predict their mean bill length, but include the original data</span></span>
<span id="cb76-2"><a href="" aria-hidden="true" tabindex="-1"></a>new_penguins <span class="sc">|&gt;</span></span>
<span id="cb76-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(</span>
<span id="cb76-4"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="fu">predict</span>(bill_length_sp_lm, <span class="at">newdata =</span> new_penguins, <span class="at">interval =</span> <span class="st">"confidence"</span>)</span>
<span id="cb76-5"><a href="" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 5
  species body_mass_g   fit   lwr   upr
  &lt;chr&gt;         &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 Adelie         4500  41.8  41.2  42.4
2 Gentoo         4500  45.3  44.8  45.9</code></pre>
</div>
</div>
</section>
<section id="confidence-intervals-3" class="slide level2">
<h2>Confidence Intervals</h2>
<ul>
<li>Confidence Intervals for model terms can also be found using <code>broom::tidy()</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## Use broom:tidy to find confidence intervals for coefficients</span></span>
<span id="cb78-2"><a href="" aria-hidden="true" tabindex="-1"></a>bill_length_sp_lm <span class="sc">|&gt;</span></span>
<span id="cb78-3"><a href="" aria-hidden="true" tabindex="-1"></a>  broom<span class="sc">::</span><span class="fu">tidy</span>(<span class="at">conf.int =</span> <span class="cn">TRUE</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 7
  term             estimate std.error statistic  p.value conf.low conf.high
  &lt;chr&gt;               &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;
1 (Intercept)      24.9      1.06         23.4  7.63e-73 22.8      27.0    
2 speciesChinstrap  9.92     0.351        28.3  5.09e-91  9.23     10.6    
3 speciesGentoo     3.56     0.486         7.32 1.78e-12  2.60      4.51   
4 body_mass_g       0.00375  0.000282     13.3  1.16e-32  0.00319   0.00430</code></pre>
</div>
</div>
</section></section>
<section>
<section id="closing-comments" class="title-slide slide level1 center">
<h1>Closing Comments</h1>

</section>
<section id="additional-plotting-comments" class="slide level2">
<h2>Additional Plotting Comments</h2>
<ul>
<li>I rarely show diagnostic plots:
<ul>
<li>For me when fitting data &amp; performing analysis</li>
<li>No need for ggplot versions</li>
</ul></li>
</ul>
</section>
<section id="section" class="slide level2">
<h2></h2>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb80-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">fitted=</span> <span class="fu">fitted</span>(bill_length_sp_lm), </span>
<span id="cb80-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">residuals =</span> <span class="fu">resid</span>(bill_length_sp_lm)</span>
<span id="cb80-4"><a href="" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span> </span>
<span id="cb80-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(fitted, residuals)) <span class="sc">+</span> </span>
<span id="cb80-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb80-7"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(</span>
<span id="cb80-8"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">se =</span> <span class="cn">FALSE</span>, <span class="at">colour =</span> <span class="st">'red'</span>, </span>
<span id="cb80-9"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">linewidth =</span> <span class="fl">0.5</span></span>
<span id="cb80-10"><a href="" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb80-11"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Residuals vs Fitted"</span>) <span class="sc">+</span></span>
<span id="cb80-12"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb80-13"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Fitted Values"</span>, <span class="at">y =</span> <span class="st">"Residuals"</span></span>
<span id="cb80-14"><a href="" aria-hidden="true" tabindex="-1"></a>  ) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="linear_regression_slides_files/figure-revealjs/unnamed-chunk-44-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div></div>
</section>
<section id="section-1" class="slide level2">
<h2></h2>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb81-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">residuals =</span> <span class="fu">rstandard</span>(bill_length_sp_lm)</span>
<span id="cb81-3"><a href="" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span> </span>
<span id="cb81-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">sample =</span> residuals)) <span class="sc">+</span> </span>
<span id="cb81-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq</span>() <span class="sc">+</span></span>
<span id="cb81-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq_line</span>() <span class="sc">+</span></span>
<span id="cb81-7"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Q-Q Residuals"</span>) <span class="sc">+</span></span>
<span id="cb81-8"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb81-9"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Theoretical Quantiles"</span>, </span>
<span id="cb81-10"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"Standardized Residuals"</span></span>
<span id="cb81-11"><a href="" aria-hidden="true" tabindex="-1"></a>  ) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="linear_regression_slides_files/figure-revealjs/unnamed-chunk-45-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div></div>
</section>
<section id="additional-plotting-packages" class="slide level2">
<h2>Additional Plotting Packages</h2>
<ul>
<li>Multiple options for great looking plots
<ul>
<li><a href="https://docs.r4photobiology.info/ggpmisc/index.html"><code>ggpmisc</code></a> for adding correlations, regression equations etc</li>
<li><a href="https://larmarange.github.io/ggstats/"><code>ggstats</code></a> for multiple fresh perspectives on coefficients</li>
<li><a href="https://indrajeetpatil.github.io/ggstatsplot/"><code>ggstatsplot</code></a> also a wide range of plotting capabilities</li>
</ul></li>
</ul>
</section>
<section id="challenges" class="slide level2">
<h2>Challenges</h2>
<div style="font-size: 90%">
<ol type="1">
<li>How could we account for <code>sex</code> in the existing <code>penguins</code> model?</li>
<li>Load the <code>pigs</code> dataset</li>
<li>Make a boxplot:
<ul>
<li><code>len</code> will be the predictor (<span class="math inline">\(y\)</span>)</li>
<li>Place <code>dose</code> on the <span class="math inline">\(x\)</span>-axis using <code>supp</code> to fill the boxes</li>
</ul></li>
<li>Find the best model for <code>len</code>, using <code>supp</code> and <code>dose</code> as the predictors</li>
<li>Check the residuals and diagnostic plots</li>
<li>Make sure you can interpret the model coefficients</li>
<li>What is the 99% confidence interval for <code>supp = "VC"</code> &amp; <code>dose = "High"</code>
<ul>
<li>What does this mean?</li>
</ul></li>
</ol>
</div>

</section></section>

    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<p><img src="assets/bodl_logo_white_background.jpg" class="slide-logo"></p>
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="linear_regression_slides_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="linear_regression_slides_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="linear_regression_slides_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="linear_regression_slides_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="linear_regression_slides_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="linear_regression_slides_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="linear_regression_slides_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="linear_regression_slides_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="linear_regression_slides_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="linear_regression_slides_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    <script type="text/javascript">
      Reveal.on('ready', event => {
        if (event.indexh === 0) {
          document.querySelector("div.has-logo > img.slide-logo").style.display = "none";
        }
      });
      Reveal.addEventListener('slidechanged', (event) => {
        if (event.indexh === 0) {
          Reveal.configure({ slideNumber: null });
          document.querySelector("div.has-logo > img.slide-logo").style.display = "none";
        }
        if (event.indexh === 1) {
          Reveal.configure({ slideNumber: 'c' });
          document.querySelector("div.has-logo > img.slide-logo").style.display = null;
        }
      });
    </script>
    

</body></html>