<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Dr Stevie Pederson">
<meta name="dcterms.date" content="2025-07-09">

<title>Linear Regression in R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="linear_regression_slides_files/libs/clipboard/clipboard.min.js"></script>
<script src="linear_regression_slides_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="linear_regression_slides_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="linear_regression_slides_files/libs/quarto-html/popper.min.js"></script>
<script src="linear_regression_slides_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="linear_regression_slides_files/libs/quarto-html/anchor.min.js"></script>
<link href="linear_regression_slides_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="linear_regression_slides_files/libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="linear_regression_slides_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="linear_regression_slides_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="linear_regression_slides_files/libs/bootstrap/bootstrap-1e1b6d7d3018d0281420745e134191b0.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  
  <script>
    (function () {
      var script = document.createElement("script");
      script.type = "text/javascript";
      script.src  = "https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js";
      document.getElementsByTagName("head")[0].appendChild(script);
    })();
  </script>


<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="bodl.scss">
<link rel="stylesheet" href="extra.css">
</head>

<body class="quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul class="collapse">
  <li><a href="#linear-regression" id="toc-linear-regression" class="nav-link active" data-scroll-target="#linear-regression">Linear Regression</a>
  <ul class="collapse">
  <li><a href="#setup" id="toc-setup" class="nav-link" data-scroll-target="#setup">Setup</a></li>
  <li><a href="#linear-regression-1" id="toc-linear-regression-1" class="nav-link" data-scroll-target="#linear-regression-1">Linear Regression</a></li>
  <li><a href="#coefficients" id="toc-coefficients" class="nav-link" data-scroll-target="#coefficients">Coefficients</a></li>
  <li><a href="#residuals" id="toc-residuals" class="nav-link" data-scroll-target="#residuals">Residuals</a></li>
  <li><a href="#assumptions-of-linear-regression" id="toc-assumptions-of-linear-regression" class="nav-link" data-scroll-target="#assumptions-of-linear-regression">Assumptions Of Linear Regression</a></li>
  <li><a href="#model-checking" id="toc-model-checking" class="nav-link" data-scroll-target="#model-checking">Model Checking</a></li>
  <li><a href="#objects-of-class-lm" id="toc-objects-of-class-lm" class="nav-link" data-scroll-target="#objects-of-class-lm">Objects Of Class <code>lm</code></a></li>
  <li><a href="#adding-terms" id="toc-adding-terms" class="nav-link" data-scroll-target="#adding-terms">Adding Terms</a></li>
  <li><a href="#model-diagnostics" id="toc-model-diagnostics" class="nav-link" data-scroll-target="#model-diagnostics">Model Diagnostics</a></li>
  <li><a href="#interpreting-the-coefficients" id="toc-interpreting-the-coefficients" class="nav-link" data-scroll-target="#interpreting-the-coefficients">Interpreting the Coefficients</a></li>
  <li><a href="#model-diagnostics-1" id="toc-model-diagnostics-1" class="nav-link" data-scroll-target="#model-diagnostics-1">Model Diagnostics</a></li>
  <li><a href="#model-selection" id="toc-model-selection" class="nav-link" data-scroll-target="#model-selection">Model Selection</a></li>
  <li><a href="#speeding-the-process-up" id="toc-speeding-the-process-up" class="nav-link" data-scroll-target="#speeding-the-process-up">Speeding The Process Up</a></li>
  <li><a href="#objects-of-class-summary.lm" id="toc-objects-of-class-summary.lm" class="nav-link" data-scroll-target="#objects-of-class-summary.lm">Objects of Class <code>summary.lm</code></a></li>
  <li><a href="#using-a-more-tidyverse-friendly-approach" id="toc-using-a-more-tidyverse-friendly-approach" class="nav-link" data-scroll-target="#using-a-more-tidyverse-friendly-approach">Using A More Tidyverse Friendly Approach</a></li>
  <li><a href="#adding-significance-stars" id="toc-adding-significance-stars" class="nav-link" data-scroll-target="#adding-significance-stars">Adding Significance Stars</a></li>
  <li><a href="#or-fitting-a-model-on-the-fly" id="toc-or-fitting-a-model-on-the-fly" class="nav-link" data-scroll-target="#or-fitting-a-model-on-the-fly">Or Fitting a Model On The Fly</a></li>
  </ul></li>
  <li><a href="#s3-method-dispatch" id="toc-s3-method-dispatch" class="nav-link" data-scroll-target="#s3-method-dispatch">S3 Method Dispatch</a>
  <ul class="collapse">
  <li><a href="#s3-method-dispatch-1" id="toc-s3-method-dispatch-1" class="nav-link" data-scroll-target="#s3-method-dispatch-1">S3 Method Dispatch</a></li>
  <li><a href="#confidence-intervals" id="toc-confidence-intervals" class="nav-link" data-scroll-target="#confidence-intervals">Confidence Intervals</a></li>
  </ul></li>
  <li><a href="#closing-comments" id="toc-closing-comments" class="nav-link" data-scroll-target="#closing-comments">Closing Comments</a>
  <ul class="collapse">
  <li><a href="#additional-plotting-comments" id="toc-additional-plotting-comments" class="nav-link" data-scroll-target="#additional-plotting-comments">Additional Plotting Comments</a></li>
  <li><a href="#section" id="toc-section" class="nav-link" data-scroll-target="#section"></a></li>
  <li><a href="#section-1" id="toc-section-1" class="nav-link" data-scroll-target="#section-1"></a></li>
  <li><a href="#additional-plotting-packages" id="toc-additional-plotting-packages" class="nav-link" data-scroll-target="#additional-plotting-packages">Additional Plotting Packages</a></li>
  <li><a href="#challenges" id="toc-challenges" class="nav-link" data-scroll-target="#challenges">Challenges</a></li>
  </ul></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="linear_regression_slides.html"><i class="bi bi-file-slides"></i>RevealJS</a></li></ul></div></nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Linear Regression in R</h1>
<p class="subtitle lead">RAdelaide 2025</p>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Dr Stevie Pederson </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            </p><div class="line-block">Black Ochre Data Labs<br>
            The Kids Research Institute Australia</div>
          <p></p>
      </div>
  </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">July 9, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="linear-regression" class="level1" data-background-color="#3d3d40">
<h1 data-background-color="#3d3d40">Linear Regression</h1>
<section id="setup" class="level2">
<h2 class="anchored" data-anchor-id="setup">Setup</h2>
<ul>
<li>Start clear R session</li>
<li>Create a blank R script: <code>LinearRegression.R</code></li>
<li>Load the key packages</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(palmerpenguins)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="linear-regression-1" class="level2">
<h2 class="anchored" data-anchor-id="linear-regression-1">Linear Regression</h2>
<p>We are trying to estimate a line with slope &amp; intercept</p>
<p><span class="math display">\[
y = ax + b
\]</span></p>
<div class="fragment">
<p>Or</p>
<p><span class="math display">\[
y = \beta_0 + \beta_1 x
\]</span></p>
</div>
<div class="incremental">
<ul class="incremental">
<li><span class="math inline">\(y\)</span> is the <em>response variable</em></li>
<li><span class="math inline">\(x\)</span> is the <em>predictor variable</em></li>
<li>Makes the most intuitive sense when both <span class="math inline">\(x\)</span> &amp; <span class="math inline">\(y\)</span> are continuous</li>
</ul>
</div>
</section>
<section id="linear-regression-2" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="linear-regression-2">Linear Regression</h2>
<p>Linear Regression always uses the <code>R</code> formula syntax</p>
<ul>
<li><code>y ~ x</code>: <code>y</code> depends on <code>x</code></li>
<li>We use the function <code>lm()</code></li>
<li>Once we have our model, we can predict <span class="math inline">\(y\)</span> based on <span class="math inline">\(x\)</span> values</li>
</ul>
<div class="fragment">
<ul>
<li>We’ll use the penguins dataset for some exploration</li>
</ul>
</div>
</section>
<section id="linear-regression-3" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="linear-regression-3">Linear Regression</h2>
<div class="cell" data-layout-align="center" data-output-location="column">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Does bill length depend on body mass?</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Plot the predictor (body mass) on `x`</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="do">## The response (bill_length) goes on `y`</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>penguins <span class="sc">|&gt;</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(body_mass_g, bill_length_mm)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="linear_regression_slides_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
<div class="notes">
<ul>
<li>There looks like a fairly clear linear relationship</li>
</ul>
</div>
</section>
<section id="linear-regression-4" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="linear-regression-4">Linear Regression</h2>
<ul>
<li>Bill length is the response variable (<span class="math inline">\(y\)</span>)</li>
<li>Body Mass is the predictor variable (<span class="math inline">\(x\)</span>)</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Fit a linear regression model and save the output as an R object</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>bill_length_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(bill_length_mm <span class="sc">~</span> body_mass_g, <span class="at">data =</span> penguins)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>bill_length_lm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = bill_length_mm ~ body_mass_g, data = penguins)

Coefficients:
(Intercept)  body_mass_g  
  26.898872     0.004051  </code></pre>
</div>
</div>
</section>
<section id="linear-regression-5" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="linear-regression-5">Linear Regression</h2>
<div class="notes">
<ul>
<li>The basic output is only minimally informative</li>
<li>Often pass to the function <code>summary()</code> to present and interpret</li>
</ul>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Use `summary` to view and interpret the fitted model</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = bill_length_mm ~ body_mass_g, data = penguins)

Residuals:
     Min       1Q   Median       3Q      Max 
-10.1251  -3.0434  -0.8089   2.0711  16.1109 

Coefficients:
             Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept) 2.690e+01  1.269e+00   21.19   &lt;2e-16 ***
body_mass_g 4.051e-03  2.967e-04   13.65   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 4.394 on 340 degrees of freedom
  (2 observations deleted due to missingness)
Multiple R-squared:  0.3542,    Adjusted R-squared:  0.3523 
F-statistic: 186.4 on 1 and 340 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
</section>
<section id="linear-regression-6" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="linear-regression-6">Linear Regression</h2>
<ol type="1">
<li>The code used is returned as <code>Call:</code></li>
<li>A brief summary of the residuals are returned</li>
<li>The fitted values are returned in the <code>Coefficients</code> element
<ul>
<li>We have the estimate of the intercept &amp; slope</li>
<li>The standard error of the estimates</li>
<li>A t-test testing <span class="math inline">\(H_0: \beta_i = 0\)</span></li>
<li>The p-value for each t-test</li>
</ul></li>
<li>Additional model summary information
<ul>
<li><span class="math inline">\(R^2\)</span> is the proportion of variance explained by the model</li>
</ul></li>
</ol>
<div class="notes">
<ul>
<li>Adjusted R-squared is the R-squared adjusted for the number of predictors</li>
<li>Mostly relevant when comparing models where predictors vary</li>
</ul>
</div>
</section>
<section id="coefficients" class="level2">
<h2 class="anchored" data-anchor-id="coefficients">Coefficients</h2>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
How do we interpret the Intercept?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="fragment">
<p>This is what the bill length would be if a penguin weighed exactly 0</p>
</div>
</div>
</div>
</div>
<div class="fragment">
<ul>
<li>We’d probably expect this to be zero but they almost never are</li>
<li>The Intercept is almost always significant</li>
<li>What does this really tell us about the relationship between bill length and weight?</li>
<li>Generally focussed on the relationship within the range of observed predictors
<ul>
<li>No guaranteed linear relationship outside of this range</li>
</ul></li>
</ul>
</div>
</section>
<section id="coefficients-1" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="coefficients-1">Coefficients</h2>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
How do we interpret the <code>body_mass_g</code> term?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="fragment">
<p>This is how much we would expect the bill length to change for <strong>every one unit increase in the predictor</strong> <br> i.e.&nbsp;for every 1<span class="math inline">\(g\)</span> increase in weight, the bill length would be expected to increase by about 0.0041mm</p>
</div>
</div>
</div>
</div>
<div class="fragment">
<ul>
<li>The <span class="math inline">\(t\)</span>-test here is highly relevant
<ul>
<li><span class="math inline">\(H_0\colon \beta_1 = 0~\)</span> Vs <span class="math inline">\(~H_A\colon \beta_1 \neq 0\)</span></li>
</ul></li>
<li>Reject <span class="math inline">\(H_0 \implies\)</span> there <em>is an association between predictor &amp; response</em></li>
</ul>
</div>
</section>
<section id="residuals" class="level2">
<h2 class="anchored" data-anchor-id="residuals">Residuals</h2>
<ul>
<li>Points never lie exactly on the regression line <span class="math inline">\(\implies\)</span> Why?</li>
</ul>
<div class="fragment">
<ul>
<li>We’re actually fitting the model</li>
</ul>
<p><span class="math display">\[
y_i = \beta_0 + \beta_1 x_i + \epsilon_i
\]</span></p>
</div>
<div class="incremental">
<ul class="incremental">
<li><span class="math inline">\(\beta_0 + \beta_1 x_i\)</span> is the exact line co-ordinate (Intercept + slope*predictor)</li>
<li><span class="math inline">\(\epsilon_i\)</span> is the the vertical difference between the observed value and the fitted value
<ul class="incremental">
<li>Known as a <em>residual</em></li>
<li>Defined as <span class="math inline">\(\epsilon_i \sim \mathcal{N}(0, \sigma)\)</span></li>
</ul></li>
</ul>
</div>
<div class="notes">
<ul>
<li>Residual means the bit left over when we subtract the predicted value from the observed</li>
</ul>
</div>
</section>
<section id="assumptions-of-linear-regression" class="level2">
<h2 class="anchored" data-anchor-id="assumptions-of-linear-regression">Assumptions Of Linear Regression</h2>
<p>Linear Regression formally has 4 key assumptions</p>
<div class="incremental">
<ol class="incremental" type="1">
<li>Linear relationship between predictor and response
<ul class="incremental">
<li>Mean of residuals is zero across the entire range</li>
</ul></li>
<li>Constant variance across the range of data (homoscedasticity)</li>
<li>Residuals are normally distributed</li>
<li>Independence of errors</li>
</ol>
</div>
<div class="incremental">
<ul class="incremental">
<li>Three of these are represented in the definition <span class="math inline">\(\epsilon_i \sim \mathcal{N}(0, \sigma)\)</span></li>
</ul>
</div>
</section>
<section id="model-checking" class="level2">
<h2 class="anchored" data-anchor-id="model-checking">Model Checking</h2>
<ul>
<li>To check our fitted model, we should check the residuals to ensure <span class="math inline">\(\epsilon_i \sim \mathcal{N}(0, \sigma)\)</span></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Check the residuals. It will ask you the following</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Hit &lt;Return&gt; to see next plot: </span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="do">## So follow the instructions and check each plot before proceeding</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="do">## There are 4 plots by default</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(bill_length_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="model-checking-1" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="model-checking-1">Model Checking</h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="linear_regression_slides_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:40%;">
<div style="font-size: 90%">
<p><br></p>
<ul>
<li>Check the zero mean of <span class="math inline">\(\mathcal{N}(0, \sigma)\)</span></li>
<li>Is this assumption satisfied across the range of the data?</li>
</ul>
</div>
</div>
</div>
</section>
<section id="model-checking-2" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="model-checking-2">Model Checking</h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="linear_regression_slides_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:40%;">
<div style="font-size: 90%">
<p><br></p>
<ul>
<li>Check the normality of <span class="math inline">\(\mathcal{N}(0, \sigma)\)</span></li>
<li>The dashed line is the expected line from a normal distribution</li>
<li>Is this assumption satisfied across the range of the data?</li>
</ul>
</div>
</div>
</div>
</section>
<section id="model-checking-3" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="model-checking-3">Model Checking</h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="linear_regression_slides_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:40%;">
<div style="font-size: 90%">
<p><br></p>
<ul>
<li>Check the constant variance of <span class="math inline">\(\mathcal{N}(0, \sigma)\)</span></li>
<li>Is this assumption satisfied across the range of the data?</li>
</ul>
</div>
</div>
</div>
</section>
<section id="model-checking-4" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="model-checking-4">Model Checking</h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="linear_regression_slides_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:40%;">
<div style="font-size: 90%">
<p><br></p>
<ul>
<li>Checks if any points are exerting excessive ‘leverage’ on the model</li>
<li>Beyond scope of today</li>
</ul>
</div>
</div>
</div>
</section>
<section id="model-checking-5" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="model-checking-5">Model Checking</h2>
<ul>
<li>All of these figures used base plotting functions</li>
<li>Stepping through can be frustrating
<ul>
<li>Especially when running an automated script</li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li>Can plot all 4 at once using a simple trick</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Define a 2x2 grid, plotting figures in a row-wise manner</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Plot the four regression diagnostic plots</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(bill_length_lm)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Reset the grid layout to be the default 1x1</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>,))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="notes">
<ul>
<li><code>par()</code> is a function to set graphical parameters</li>
<li><code>mfrow</code> means ‘Multi-Figures By Row’</li>
<li>c(2, 2,) means plot in a 2x2 grid</li>
<li>I rarely show these so no need for ggplot versions</li>
</ul>
</div>
</section>
<section id="objects-of-class-lm" class="level2">
<h2 class="anchored" data-anchor-id="objects-of-class-lm">Objects Of Class <code>lm</code></h2>
<ul>
<li>The linear model we fitted produced an object of class <code>lm</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(bill_length_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "lm"</code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>This is also a list</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Inspect the actual object</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(bill_length_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="objects-of-class-lm-1" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="objects-of-class-lm-1">Objects Of Class <code>lm</code></h2>
<ul>
<li>We can use the list structure to inspect the residuals manually</li>
</ul>
<div class="cell" data-layout-align="center" data-output-location="column">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Plot a histogram of the residuals</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(bill_length_lm<span class="sc">$</span>residuals)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="linear_regression_slides_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="objects-of-class-lm-2" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="objects-of-class-lm-2">Objects Of Class <code>lm</code></h2>
<ul>
<li>We could even use the Shapiro-Wilk test for normality</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">shapiro.test</span>(bill_length_lm<span class="sc">$</span>residuals)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Shapiro-Wilk normality test

data:  bill_length_lm$residuals
W = 0.95439, p-value = 8.217e-09</code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>How can we interpret all of this?</li>
<li>Maybe there’s a better model</li>
</ul>
</div>
</section>
<section id="adding-terms" class="level2">
<h2 class="anchored" data-anchor-id="adding-terms">Adding Terms</h2>
<div class="cell" data-layout-align="center" data-output-location="column">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Does bill length depend on body mass?</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Plot the predictor (body mass) on `x`</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="do">## The response (bill_length) goes on `y`</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Does each species need it's own model</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>penguins <span class="sc">|&gt;</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>      body_mass_g, bill_length_mm, </span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour =</span> species</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="linear_regression_slides_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
<div class="notes">
<ul>
<li>Do we think these slopes are the same?</li>
<li>Do we think the intercepts are the same?</li>
</ul>
</div>
</section>
<section id="adding-terms-1" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="adding-terms-1">Adding Terms</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Include the species in the model</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="do">## NB: This will fit a separate intercept for each species</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>bill_length_sp_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(bill_length_mm <span class="sc">~</span> species <span class="sc">+</span> body_mass_g, <span class="at">data =</span> penguins) </span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_sp_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = bill_length_mm ~ species + body_mass_g, data = penguins)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.8129 -1.6718  0.1336  1.4720  9.2902 

Coefficients:
                  Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)      2.492e+01  1.063e+00  23.443  &lt; 2e-16 ***
speciesChinstrap 9.921e+00  3.511e-01  28.258  &lt; 2e-16 ***
speciesGentoo    3.558e+00  4.858e-01   7.324 1.78e-12 ***
body_mass_g      3.749e-03  2.823e-04  13.276  &lt; 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 2.403 on 338 degrees of freedom
  (2 observations deleted due to missingness)
Multiple R-squared:  0.808, Adjusted R-squared:  0.8063 
F-statistic:   474 on 3 and 338 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<div class="notes">
<ul>
<li>There is already a large increase in R-squared so this model may fit better</li>
</ul>
</div>
</section>
<section id="adding-terms-2" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="adding-terms-2">Adding Terms</h2>
<div class="cell" data-layout-align="center" data-output-location="column">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Check the residuals after </span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="do">## including a separate intercept</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="do">## for species</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(bill_length_sp_lm)</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="linear_regression_slides_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="model-diagnostics" class="level2">
<h2 class="anchored" data-anchor-id="model-diagnostics">Model Diagnostics</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Check the histogram of residuals</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(bill_length_sp_lm<span class="sc">$</span>residuals, <span class="at">breaks =</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Perform the Shapiro-Wilk test for Normality</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">shapiro.test</span>(bill_length_sp_lm<span class="sc">$</span>residuals)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Shapiro-Wilk normality test

data:  bill_length_sp_lm$residuals
W = 0.99317, p-value = 0.123</code></pre>
</div>
</div>
</section>
<section id="interpreting-the-coefficients" class="level2">
<h2 class="anchored" data-anchor-id="interpreting-the-coefficients">Interpreting the Coefficients</h2>
<ul>
<li>Now we’re happier with the model <span class="math inline">\(\rightarrow\)</span> what do the coefficients mean?</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code>                     Estimate   Std. Error   t value     Pr(&gt;|t|)
(Intercept)      24.919470977 1.0630034684 23.442511 7.632983e-73
speciesChinstrap  9.920884113 0.3510790185 28.258265 5.093822e-91
speciesGentoo     3.557977539 0.4857896978  7.324111 1.776921e-12
body_mass_g       0.003748497 0.0002823439 13.276352 1.158990e-32</code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>By default, the primary Intercept is the first factor level in <code>species</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(penguins<span class="sc">$</span>species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Adelie"    "Chinstrap" "Gentoo"   </code></pre>
</div>
</div>
</div>
</section>
<section id="interpreting-the-coefficients-1" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="interpreting-the-coefficients-1">Interpreting the Coefficients</h2>
<ul>
<li>Now we’re happier with the model <span class="math inline">\(\rightarrow\)</span> what do the coefficients mean?</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code>                     Estimate   Std. Error   t value     Pr(&gt;|t|)
(Intercept)      24.919470977 1.0630034684 23.442511 7.632983e-73
speciesChinstrap  9.920884113 0.3510790185 28.258265 5.093822e-91
speciesGentoo     3.557977539 0.4857896978  7.324111 1.776921e-12
body_mass_g       0.003748497 0.0002823439 13.276352 1.158990e-32</code></pre>
</div>
</div>
<ul>
<li>The baseline intercept is for Adelie penguins</li>
<li>Additional intercept terms are the differences between the baseline and each species
<ul>
<li>Does this check-out in our initial plot of the data?</li>
</ul></li>
<li>They all appear significant when checking each <span class="math inline">\(H_0\)</span></li>
</ul>
</section>
<section id="adding-terms-3" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="adding-terms-3">Adding Terms</h2>
<ul>
<li>Do we think each species may have a different relationship between mass and bill length?
<ul>
<li>Do we need to fit a separate slope for each species?</li>
</ul></li>
<li>This is done in <code>R</code> using an “interaction term” separating terms by <code>:</code>
<ul>
<li><code>species:body_mass_g</code></li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>bill_length_int_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  bill_length_mm <span class="sc">~</span> species <span class="sc">+</span> body_mass_g <span class="sc">+</span> species<span class="sc">:</span>body_mass_g, </span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> penguins</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>) </span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_int_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = bill_length_mm ~ species + body_mass_g + species:body_mass_g, 
    data = penguins)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.4208 -1.6461  0.0919  1.4718  9.3138 

Coefficients:
                               Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)                  26.9941391  1.5926015  16.950  &lt; 2e-16 ***
speciesChinstrap              5.1800537  3.2746719   1.582    0.115    
speciesGentoo                -0.2545907  2.7138655  -0.094    0.925    
body_mass_g                   0.0031879  0.0004271   7.464 7.27e-13 ***
speciesChinstrap:body_mass_g  0.0012748  0.0008740   1.459    0.146    
speciesGentoo:body_mass_g     0.0009030  0.0006066   1.489    0.138    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 2.399 on 336 degrees of freedom
  (2 observations deleted due to missingness)
Multiple R-squared:  0.8098,    Adjusted R-squared:  0.807 
F-statistic: 286.1 on 5 and 336 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<div class="notes">
<ul>
<li>Now what do we think?</li>
</ul>
</div>
</section>
<section id="interpreting-the-coefficients-2" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="interpreting-the-coefficients-2">Interpreting the Coefficients</h2>
<ul>
<li>The baseline terms for the Intercept and <code>body_mass_g</code> are now <strong>both</strong> for Adelie</li>
<li>Differences in slope for each species are provided as <code>speciesChinstrap:body_mass_g</code> and <code>speciesGentoo:body_mass_g</code></li>
</ul>
<div class="incremental">
<ul class="incremental">
<li><p>The regression line for Adelie is y = 26.99 + 0.0032*<code>body_mass_g</code></p></li>
<li><p>For Chinstrap: y = (26.99+5.18) + (0.0032+0.0013)*body_mass_g`</p></li>
</ul>
</div>
</section>
<section id="interpreting-the-coefficients-3" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="interpreting-the-coefficients-3">Interpreting the Coefficients</h2>
<ul>
<li>The model matrix <span class="math inline">\(X\)</span> is a key part of understanding statistics</li>
<li>The model coefficients (<span class="math inline">\(\hat{\beta}\)</span>) are actually estimated by performing linear algebra on this</li>
<li>The least squares solution is <span class="math inline">\(\hat{\beta} = (X^TX)^{-1}X^Ty\)</span></li>
</ul>
<div class="fragment">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> penguins<span class="sc">$</span>bill_length_mm[<span class="sc">!</span><span class="fu">is.na</span>(penguins<span class="sc">$</span>body_mass_g)]</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> species <span class="sc">+</span> body_mass_g <span class="sc">+</span> species<span class="sc">:</span>body_mass_g, <span class="at">data =</span> penguins)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="interpreting-the-coefficients-4" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="interpreting-the-coefficients-4">Interpreting the Coefficients</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  (Intercept) speciesChinstrap speciesGentoo body_mass_g
1           1                0             0        3750
2           1                0             0        3800
3           1                0             0        3250
5           1                0             0        3450
6           1                0             0        3650
7           1                0             0        3625
  speciesChinstrap:body_mass_g speciesGentoo:body_mass_g
1                            0                         0
2                            0                         0
3                            0                         0
5                            0                         0
6                            0                         0
7                            0                         0</code></pre>
</div>
</div>
</section>
<section id="interpreting-the-coefficients-5" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="interpreting-the-coefficients-5">Interpreting the Coefficients</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="do">## No need to save this, I'm just demonstrating a point</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="do">## The inverse of a matrix is found using `solve()`</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Matrix multiplication is performed using %*%</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="do">## The transpose of a matrix is found using `t()`</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>beta_hat <span class="ot">&lt;-</span> <span class="fu">solve</span>(<span class="fu">t</span>(X) <span class="sc">%*%</span> X) <span class="sc">%*%</span> <span class="fu">t</span>(X) <span class="sc">%*%</span> y</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="fu">cbind</span>(beta_hat, <span class="fu">coef</span>(bill_length_int_lm))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                      [,1]          [,2]
(Intercept)                  26.9941391366 26.9941391367
speciesChinstrap              5.1800537287  5.1800537287
speciesGentoo                -0.2545906615 -0.2545906615
body_mass_g                   0.0031878758  0.0031878758
speciesChinstrap:body_mass_g  0.0012748183  0.0012748183
speciesGentoo:body_mass_g     0.0009029956  0.0009029956</code></pre>
</div>
</div>
</section>
<section id="interpreting-the-coefficients-6" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="interpreting-the-coefficients-6">Interpreting the Coefficients</h2>
<div style="font-size: 85%">
<ul>
<li>An excellent primer on model matrices is <a href="https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html">here</a><a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></li>
<li>Conventional statistics always sets a baseline level for the intercept
<ul>
<li>Performs a <span class="math inline">\(t\)</span>-test automatically for differences</li>
<li>Really quite convenient if less obvious</li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li>We could remove the common intercept using <code>~ 0 + species + body_mass_g.</code>
<ul>
<li>Would return a separate intercept by removing the common (baseline) level</li>
<li>No real advantage except non-statisticians like it</li>
</ul></li>
<li>No statistical evidence provided for differences in intercepts<br><span class="math inline">\(\implies\)</span> have to test ourselves 😢</li>
</ul>
</div>
</div>
</section>
<section id="model-diagnostics-1" class="level2">
<h2 class="anchored" data-anchor-id="model-diagnostics-1">Model Diagnostics</h2>
<div class="cell" data-layout-align="center" data-output-location="column">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Check the residuals after </span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="do">## including a separate intercept</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="do">## and slope for species</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(bill_length_int_lm)</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="linear_regression_slides_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="model-selection" class="level2">
<h2 class="anchored" data-anchor-id="model-selection">Model Selection</h2>
<ul>
<li>How do we decide on the best model?</li>
<li>A common technique is Analysis of Variance (ANOVA)</li>
<li>Classic ANOVA checks importance of each term within a model</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Run a separate ANOVA on each model</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(bill_length_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: bill_length_mm
             Df Sum Sq Mean Sq F value    Pr(&gt;F)    
body_mass_g   1 3599.7  3599.7  186.44 &lt; 2.2e-16 ***
Residuals   340 6564.5    19.3                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(bill_length_sp_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: bill_length_mm
             Df Sum Sq Mean Sq F value    Pr(&gt;F)    
species       2 7194.3  3597.2  622.88 &lt; 2.2e-16 ***
body_mass_g   1 1017.9  1017.9  176.26 &lt; 2.2e-16 ***
Residuals   338 1952.0     5.8                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(bill_length_int_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: bill_length_mm
                     Df Sum Sq Mean Sq  F value Pr(&gt;F)    
species               2 7194.3  3597.2 625.1488 &lt;2e-16 ***
body_mass_g           1 1017.9  1017.9 176.9039 &lt;2e-16 ***
species:body_mass_g   2   18.6     9.3   1.6159 0.2003    
Residuals           336 1933.4     5.8                    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>Does this give any clue as to the best model?</li>
</ul>
</div>
</section>
<section id="model-selection-1" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="model-selection-1">Model Selection</h2>
<ul>
<li>We have progressively added terms to each model</li>
<li>Can use ANOVA to compare suitability of each model</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Compare the models using ANOVA</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="do">## If a model is a significant improvement, the p-value from the </span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="do">## F test will be clearly significant</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>  bill_length_lm,</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>  bill_length_sp_lm,</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>  bill_length_int_lm</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Model 1: bill_length_mm ~ body_mass_g
Model 2: bill_length_mm ~ species + body_mass_g
Model 3: bill_length_mm ~ species + body_mass_g + species:body_mass_g
  Res.Df    RSS Df Sum of Sq        F Pr(&gt;F)    
1    340 6564.5                                 
2    338 1952.0  2    4612.5 400.8045 &lt;2e-16 ***
3    336 1933.4  2      18.6   1.6159 0.2003    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>It looks like the separate slopes are not an improvement</li>
<li>The separate intercepts <strong>are</strong> an improvement</li>
</ul>
</div>
</section>
<section id="speeding-the-process-up" class="level2">
<h2 class="anchored" data-anchor-id="speeding-the-process-up">Speeding The Process Up</h2>
<ul>
<li>This was a careful breakdown of finding the best model</li>
<li>We can partially automate this and use some shortcuts</li>
</ul>
<div class="fragment">
<ul>
<li>The shorthand for an interaction term with separate intercepts is <code>*</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Refit the model with an interaction term using the shorthand code</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>bill_length_int_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>  bill_length_mm <span class="sc">~</span> species <span class="sc">*</span> body_mass_g, <span class="at">data =</span> penguins</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_int_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = bill_length_mm ~ species * body_mass_g, data = penguins)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.4208 -1.6461  0.0919  1.4718  9.3138 

Coefficients:
                               Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)                  26.9941391  1.5926015  16.950  &lt; 2e-16 ***
speciesChinstrap              5.1800537  3.2746719   1.582    0.115    
speciesGentoo                -0.2545907  2.7138655  -0.094    0.925    
body_mass_g                   0.0031879  0.0004271   7.464 7.27e-13 ***
speciesChinstrap:body_mass_g  0.0012748  0.0008740   1.459    0.146    
speciesGentoo:body_mass_g     0.0009030  0.0006066   1.489    0.138    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 2.399 on 336 degrees of freedom
  (2 observations deleted due to missingness)
Multiple R-squared:  0.8098,    Adjusted R-squared:  0.807 
F-statistic: 286.1 on 5 and 336 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
</div>
<div class="fragment">
<ul>
<li>Alternatively, all terms can be placed inside brackets &amp; raised to a power
<ul>
<li><code>(species + body_mass_g)^2</code> would give two-way interactions</li>
</ul></li>
</ul>
</div>
</section>
<section id="speeding-the-process-up-1" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="speeding-the-process-up-1">Speeding The Process Up</h2>
<div class="notes">
<ul>
<li>The AIC is a function of the number of model terms and the log likelihood function</li>
<li>Beyond the scope of today</li>
<li>Doesn’t perform well with highly correlated predictors</li>
</ul>
</div>
<ul>
<li>After specifying a ‘full’ model <span class="math inline">\(\implies\)</span> use <code>step()</code> to remove redundant terms
<ul>
<li>Removes terms in a step-wise manner</li>
<li>Uses Akaike’s Information Criterion (AIC) to determine optimal model</li>
<li>Finds model with lowest AIC</li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">step</span>(bill_length_int_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Start:  AIC=604.42
bill_length_mm ~ species * body_mass_g

                      Df Sum of Sq    RSS    AIC
- species:body_mass_g  2    18.596 1952.0 603.69
&lt;none&gt;                             1933.4 604.42

Step:  AIC=603.69
bill_length_mm ~ species + body_mass_g

              Df Sum of Sq    RSS     AIC
&lt;none&gt;                     1952.0  603.69
- body_mass_g  1    1017.9 2969.9  745.22
- species      2    4612.5 6564.5 1014.48</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = bill_length_mm ~ species + body_mass_g, data = penguins)

Coefficients:
     (Intercept)  speciesChinstrap     speciesGentoo       body_mass_g  
       24.919471          9.920884          3.557978          0.003748  </code></pre>
</div>
</div>
</section>
<section id="objects-of-class-summary.lm" class="level2">
<h2 class="anchored" data-anchor-id="objects-of-class-summary.lm">Objects of Class <code>summary.lm</code></h2>
<ul>
<li>The coefficients element of the basic <code>lm</code> object only had the fitted values
<ul>
<li>Not the std errors, t-tests or p-values</li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Extract the coefficients directly from the linear model</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>bill_length_sp_lm<span class="sc">$</span>coefficients</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     (Intercept) speciesChinstrap    speciesGentoo      body_mass_g 
    24.919470977      9.920884113      3.557977539      0.003748497 </code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>These values are produced by the function <code>summary()</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Extract the coefficients from the summary object</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_sp_lm)<span class="sc">$</span>coefficients</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                     Estimate   Std. Error   t value     Pr(&gt;|t|)
(Intercept)      24.919470977 1.0630034684 23.442511 7.632983e-73
speciesChinstrap  9.920884113 0.3510790185 28.258265 5.093822e-91
speciesGentoo     3.557977539 0.4857896978  7.324111 1.776921e-12
body_mass_g       0.003748497 0.0002823439 13.276352 1.158990e-32</code></pre>
</div>
</div>
</div>
</section>
<section id="objects-of-class-summary.lm-1" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="objects-of-class-summary.lm-1">Objects of Class <code>summary.lm</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Check the class of the output from summary</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_sp_lm) <span class="sc">|&gt;</span> <span class="fu">class</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "summary.lm"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Prove conclusively that it is really a list</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="do">## with a class attribute stuck on it</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_sp_lm) <span class="sc">|&gt;</span> <span class="fu">is.list</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="do">## See what attributes it has</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_sp_lm) <span class="sc">|&gt;</span> <span class="fu">attributes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$names
 [1] "call"          "terms"         "residuals"     "coefficients" 
 [5] "aliased"       "sigma"         "df"            "r.squared"    
 [9] "adj.r.squared" "fstatistic"    "cov.unscaled"  "na.action"    

$class
[1] "summary.lm"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Check the full details</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bill_length_sp_lm) <span class="sc">|&gt;</span> <span class="fu">glimpse</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 12
 $ call         : language lm(formula = bill_length_mm ~ species + body_mass_g, data = penguins)
 $ terms        :Classes 'terms', 'formula'  language bill_length_mm ~ species + body_mass_g
  .. ..- attr(*, "variables")= language list(bill_length_mm, species, body_mass_g)
  .. ..- attr(*, "factors")= int [1:3, 1:2] 0 1 0 0 0 1
  .. .. ..- attr(*, "dimnames")=List of 2
  .. ..- attr(*, "term.labels")= chr [1:2] "species" "body_mass_g"
  .. ..- attr(*, "order")= int [1:2] 1 1
  .. ..- attr(*, "intercept")= int 1
  .. ..- attr(*, "response")= int 1
  .. ..- attr(*, ".Environment")=&lt;environment: R_GlobalEnv&gt; 
  .. ..- attr(*, "predvars")= language list(bill_length_mm, species, body_mass_g)
  .. ..- attr(*, "dataClasses")= Named chr [1:3] "numeric" "factor" "numeric"
  .. .. ..- attr(*, "names")= chr [1:3] "bill_length_mm" "species" "body_mass_g"
 $ residuals    : Named num [1:342] 0.124 0.336 3.198 -1.152 0.699 ...
  ..- attr(*, "names")= chr [1:342] "1" "2" "3" "5" ...
 $ coefficients : num [1:4, 1:4] 24.91947 9.92088 3.55798 0.00375 1.063 ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : chr [1:4] "(Intercept)" "speciesChinstrap" "speciesGentoo" "body_mass_g"
  .. ..$ : chr [1:4] "Estimate" "Std. Error" "t value" "Pr(&gt;|t|)"
 $ aliased      : Named logi [1:4] FALSE FALSE FALSE FALSE
  ..- attr(*, "names")= chr [1:4] "(Intercept)" "speciesChinstrap" "speciesGentoo" "body_mass_g"
 $ sigma        : num 2.4
 $ df           : int [1:3] 4 338 4
 $ r.squared    : num 0.808
 $ adj.r.squared: num 0.806
 $ fstatistic   : Named num [1:3] 474 3 338
  ..- attr(*, "names")= chr [1:3] "value" "numdf" "dendf"
 $ cov.unscaled : num [1:4, 1:4] 1.96e-01 -4.97e-03 6.36e-02 -5.11e-05 -4.97e-03 ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : chr [1:4] "(Intercept)" "speciesChinstrap" "speciesGentoo" "body_mass_g"
  .. ..$ : chr [1:4] "(Intercept)" "speciesChinstrap" "speciesGentoo" "body_mass_g"
 $ na.action    : 'omit' Named int [1:2] 4 272
  ..- attr(*, "names")= chr [1:2] "4" "272"
 - attr(*, "class")= chr "summary.lm"</code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>The full complexity of the object is mostly irrelevant</li>
</ul>
</div>
</section>
<section id="using-a-more-tidyverse-friendly-approach" class="level2">
<h2 class="anchored" data-anchor-id="using-a-more-tidyverse-friendly-approach">Using A More Tidyverse Friendly Approach</h2>
<ul>
<li>The function <code>tidy()</code> from the package <code>broom</code> is a catch-all function
<ul>
<li>Will return a tibble</li>
<li>Returns the same from <code>lm</code> and <code>summary.lm</code> objects</li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>bill_length_sp_lm <span class="sc">|&gt;</span> broom<span class="sc">::</span><span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  term             estimate std.error statistic  p.value
  &lt;chr&gt;               &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)      24.9      1.06         23.4  7.63e-73
2 speciesChinstrap  9.92     0.351        28.3  5.09e-91
3 speciesGentoo     3.56     0.486         7.32 1.78e-12
4 body_mass_g       0.00375  0.000282     13.3  1.16e-32</code></pre>
</div>
</div>
</section>
<section id="adding-significance-stars" class="level2">
<h2 class="anchored" data-anchor-id="adding-significance-stars">Adding Significance Stars</h2>
<ul>
<li>The easiest way for me as a <code>case_when()</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>bill_length_sp_lm <span class="sc">|&gt;</span> </span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>  broom<span class="sc">::</span><span class="fu">tidy</span>() <span class="sc">|&gt;</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">stars =</span> <span class="fu">case_when</span>(</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.001</span> <span class="sc">~</span> <span class="st">"***"</span>,</span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.01</span> <span class="sc">~</span> <span class="st">"**"</span>,</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">~</span> <span class="st">"*"</span>,</span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.1</span> <span class="sc">~</span> <span class="st">"."</span>,</span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a>      <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">""</span></span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 6
  term             estimate std.error statistic  p.value stars
  &lt;chr&gt;               &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;
1 (Intercept)      24.9      1.06         23.4  7.63e-73 ***  
2 speciesChinstrap  9.92     0.351        28.3  5.09e-91 ***  
3 speciesGentoo     3.56     0.486         7.32 1.78e-12 ***  
4 body_mass_g       0.00375  0.000282     13.3  1.16e-32 ***  </code></pre>
</div>
</div>
</section>
<section id="or-fitting-a-model-on-the-fly" class="level2">
<h2 class="anchored" data-anchor-id="or-fitting-a-model-on-the-fly">Or Fitting a Model On The Fly</h2>
<ul>
<li>Obviously no room for checking model diagnostics</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>penguins <span class="sc">|&gt;</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Piped data can be recalled using `_`</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lm</span>( bill_length_mm <span class="sc">~</span> species <span class="sc">*</span> body_mass_g, <span class="at">data =</span> _) <span class="sc">|&gt;</span> </span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">step</span>() <span class="sc">|&gt;</span> <span class="co"># Fit the best model using the AIC</span></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>  broom<span class="sc">::</span><span class="fu">tidy</span>() <span class="sc">|&gt;</span> <span class="co"># Turn the output into a tibble &amp; add stars</span></span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">stars =</span> <span class="fu">case_when</span>(</span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.001</span> <span class="sc">~</span> <span class="st">"***"</span>,</span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.01</span> <span class="sc">~</span> <span class="st">"**"</span>,</span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">~</span> <span class="st">"*"</span>,</span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.1</span> <span class="sc">~</span> <span class="st">"."</span>,</span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a>      <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">""</span></span>
<span id="cb63-13"><a href="#cb63-13" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb63-14"><a href="#cb63-14" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Start:  AIC=604.42
bill_length_mm ~ species * body_mass_g

                      Df Sum of Sq    RSS    AIC
- species:body_mass_g  2    18.596 1952.0 603.69
&lt;none&gt;                             1933.4 604.42

Step:  AIC=603.69
bill_length_mm ~ species + body_mass_g

              Df Sum of Sq    RSS     AIC
&lt;none&gt;                     1952.0  603.69
- body_mass_g  1    1017.9 2969.9  745.22
- species      2    4612.5 6564.5 1014.48</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 6
  term             estimate std.error statistic  p.value stars
  &lt;chr&gt;               &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;
1 (Intercept)      24.9      1.06         23.4  7.63e-73 ***  
2 speciesChinstrap  9.92     0.351        28.3  5.09e-91 ***  
3 speciesGentoo     3.56     0.486         7.32 1.78e-12 ***  
4 body_mass_g       0.00375  0.000282     13.3  1.16e-32 ***  </code></pre>
</div>
</div>
</section>
</section>
<section id="s3-method-dispatch" class="level1">
<h1>S3 Method Dispatch</h1>
<section id="s3-method-dispatch-1" class="level2">
<h2 class="anchored" data-anchor-id="s3-method-dispatch-1">S3 Method Dispatch</h2>
<ul>
<li>Objects with class <code>lm</code> and <code>summary.lm</code> are <code>S3</code> objects</li>
<li>Very informal class structure in <code>R</code></li>
<li>Easy to work with <span class="math inline">\(\implies\)</span> easy to break</li>
</ul>
<div class="fragment">
<ul>
<li>When we printed these objects <span class="math inline">\(\implies\)</span> <code>print.lm()</code> or <code>print.summary.lm()</code></li>
<li>Likewise when we plotted the <code>lm</code> object <span class="math inline">\(\implies\)</span> <code>plot.lm()</code></li>
</ul>
</div>
<div class="fragment">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>?plot.lm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="s3-method-dispatch-2" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="s3-method-dispatch-2">S3 Method Dispatch</h2>
<ul>
<li>If we only want a subset of figures</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  bill_length_sp_lm, <span class="at">which =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>),</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">caption =</span> <span class="fu">list</span>(</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Separate Intercepts: Residuals vs Fitted"</span>, <span class="cn">NULL</span>,</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Separate Intercepts: Scale-Location"</span></span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="linear_regression_slides_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="768"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="s3-method-dispatch-3" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="s3-method-dispatch-3">S3 Method Dispatch</h2>
<ul>
<li>We can also check what we pass to <code>broom::tidy()</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>?broom<span class="sc">::</span>tidy.lm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<ul>
<li><code>broom::tidy()</code> has multiple versions for objects of different classes
<ul>
<li>e.g.&nbsp;<code>broom::tidy.prcomp()</code> for PCA results</li>
</ul></li>
</ul>
</div>
</section>
<section id="s3-method-dispatch-4" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="s3-method-dispatch-4">S3 Method Dispatch</h2>
<ul>
<li><code>S3</code> objects can be easily broken</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>broken_lm <span class="ot">&lt;-</span> bill_length_sp_lm</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(broken_lm) <span class="ot">&lt;-</span> <span class="st">"htest"</span></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>broken_lm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>


data:  </code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li><code>R</code> looked for the <code>print.htest</code> method</li>
<li>The structure didn’t match what was expected <span class="math inline">\(\implies\)</span> nonsense output</li>
</ul>
</div>
</section>
<section id="confidence-intervals" class="level2">
<h2 class="anchored" data-anchor-id="confidence-intervals">Confidence Intervals</h2>
<ul>
<li>For confidence or prediction intervals, we can use <code>predict.lm()</code></li>
</ul>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
What’s the difference between a confidence and prediction interval
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="fragment">
<ul>
<li>95% Confidence Intervals: The mean will be in the given interval 95% of the time</li>
<li>95% Prediction Intervals: An observation will be in the given interval 95% of the time</li>
</ul>
</div>
</div>
</div>
</div>
<div class="fragment">
<ul>
<li>We will need a new data frame to make the predictions about</li>
</ul>
</div>
</section>
<section id="confidence-intervals-1" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="confidence-intervals-1">Confidence Intervals</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Create some new penguins</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>new_penguins <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">species =</span> <span class="fu">c</span>(<span class="st">"Adelie"</span>, <span class="st">"Gentoo"</span>), </span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">body_mass_g =</span> <span class="dv">4500</span></span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Predict their mean bill length</span></span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a><span class="fu">predict</span>(bill_length_sp_lm, <span class="at">newdata =</span> new_penguins, <span class="at">interval =</span> <span class="st">"confidence"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       fit      lwr      upr
1 41.78771 41.20030 42.37512
2 45.34568 44.81277 45.87860</code></pre>
</div>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Predict the range a new observation may lie in</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a><span class="fu">predict</span>(bill_length_sp_lm, <span class="at">newdata =</span> new_penguins, <span class="at">interval =</span> <span class="st">"prediction"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       fit      lwr      upr
1 41.78771 37.02436 46.55105
2 45.34568 40.58876 50.10261</code></pre>
</div>
</div>
</section>
<section id="confidence-intervals-2" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="confidence-intervals-2">Confidence Intervals</h2>
<ul>
<li>We may wish to include our new penguins in these results</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Predict their mean bill length, but include the original data</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>new_penguins <span class="sc">|&gt;</span></span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(</span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">predict</span>(bill_length_sp_lm, <span class="at">newdata =</span> new_penguins, <span class="at">interval =</span> <span class="st">"confidence"</span>)</span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 5
  species body_mass_g   fit   lwr   upr
  &lt;chr&gt;         &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 Adelie         4500  41.8  41.2  42.4
2 Gentoo         4500  45.3  44.8  45.9</code></pre>
</div>
</div>
</section>
<section id="confidence-intervals-3" class="level2 slide-only unlisted">
<h2 class="slide-only unlisted anchored" data-anchor-id="confidence-intervals-3">Confidence Intervals</h2>
<ul>
<li>Confidence Intervals for model terms can also be found using <code>broom::tidy()</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Use broom:tidy to find confidence intervals for coefficients</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>bill_length_sp_lm <span class="sc">|&gt;</span></span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>  broom<span class="sc">::</span><span class="fu">tidy</span>(<span class="at">conf.int =</span> <span class="cn">TRUE</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 7
  term             estimate std.error statistic  p.value conf.low conf.high
  &lt;chr&gt;               &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;
1 (Intercept)      24.9      1.06         23.4  7.63e-73 22.8      27.0    
2 speciesChinstrap  9.92     0.351        28.3  5.09e-91  9.23     10.6    
3 speciesGentoo     3.56     0.486         7.32 1.78e-12  2.60      4.51   
4 body_mass_g       0.00375  0.000282     13.3  1.16e-32  0.00319   0.00430</code></pre>
</div>
</div>
</section>
</section>
<section id="closing-comments" class="level1">
<h1>Closing Comments</h1>
<section id="additional-plotting-comments" class="level2">
<h2 class="anchored" data-anchor-id="additional-plotting-comments">Additional Plotting Comments</h2>
<ul>
<li>I rarely show diagnostic plots:
<ul>
<li>For me when fitting data &amp; performing analysis</li>
<li>No need for ggplot versions</li>
</ul></li>
</ul>
</section>
<section id="section" class="level2">
<h2 class="anchored" data-anchor-id="section"></h2>
<div class="cell" data-layout-align="center" data-output-location="column">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">fitted=</span> <span class="fu">fitted</span>(bill_length_sp_lm), </span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">residuals =</span> <span class="fu">resid</span>(bill_length_sp_lm)</span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span> </span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(fitted, residuals)) <span class="sc">+</span> </span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(</span>
<span id="cb80-8"><a href="#cb80-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">se =</span> <span class="cn">FALSE</span>, <span class="at">colour =</span> <span class="st">'red'</span>, </span>
<span id="cb80-9"><a href="#cb80-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">linewidth =</span> <span class="fl">0.5</span></span>
<span id="cb80-10"><a href="#cb80-10" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb80-11"><a href="#cb80-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Residuals vs Fitted"</span>) <span class="sc">+</span></span>
<span id="cb80-12"><a href="#cb80-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb80-13"><a href="#cb80-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Fitted Values"</span>, <span class="at">y =</span> <span class="st">"Residuals"</span></span>
<span id="cb80-14"><a href="#cb80-14" aria-hidden="true" tabindex="-1"></a>  ) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="linear_regression_slides_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="section-1" class="level2">
<h2 class="anchored" data-anchor-id="section-1"></h2>
<div class="cell" data-layout-align="center" data-output-location="column">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">residuals =</span> <span class="fu">rstandard</span>(bill_length_sp_lm)</span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span> </span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">sample =</span> residuals)) <span class="sc">+</span> </span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq</span>() <span class="sc">+</span></span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq_line</span>() <span class="sc">+</span></span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Q-Q Residuals"</span>) <span class="sc">+</span></span>
<span id="cb81-8"><a href="#cb81-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb81-9"><a href="#cb81-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Theoretical Quantiles"</span>, </span>
<span id="cb81-10"><a href="#cb81-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"Standardized Residuals"</span></span>
<span id="cb81-11"><a href="#cb81-11" aria-hidden="true" tabindex="-1"></a>  ) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="linear_regression_slides_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="additional-plotting-packages" class="level2">
<h2 class="anchored" data-anchor-id="additional-plotting-packages">Additional Plotting Packages</h2>
<ul>
<li>Multiple options for great looking plots
<ul>
<li><a href="https://docs.r4photobiology.info/ggpmisc/index.html"><code>ggpmisc</code></a> for adding correlations, regression equations etc</li>
<li><a href="https://larmarange.github.io/ggstats/"><code>ggstats</code></a> for multiple fresh perspectives on coefficients</li>
<li><a href="https://indrajeetpatil.github.io/ggstatsplot/"><code>ggstatsplot</code></a> also a wide range of plotting capabilities</li>
</ul></li>
</ul>
</section>
<section id="challenges" class="level2">
<h2 class="anchored" data-anchor-id="challenges">Challenges</h2>
<div style="font-size: 90%">
<ol type="1">
<li>How could we account for <code>sex</code> in the existing <code>penguins</code> model?</li>
<li>Load the <code>pigs</code> dataset</li>
<li>Make a boxplot:
<ul>
<li><code>len</code> will be the predictor (<span class="math inline">\(y\)</span>)</li>
<li>Place <code>dose</code> on the <span class="math inline">\(x\)</span>-axis using <code>supp</code> to fill the boxes</li>
</ul></li>
<li>Find the best model for <code>len</code>, using <code>supp</code> and <code>dose</code> as the predictors</li>
<li>Check the residuals and diagnostic plots</li>
<li>Make sure you can interpret the model coefficients</li>
<li>What is the 99% confidence interval for <code>supp = "VC"</code> &amp; <code>dose = "High"</code>
<ul>
<li>What does this mean?</li>
</ul></li>
</ol>
</div>
</section>
</section>


<div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->
<script type="text/javascript">
  Reveal.on('ready', event => {
    if (event.indexh === 0) {
      document.querySelector("div.has-logo > img.slide-logo").style.display = "none";
    }
  });
  Reveal.addEventListener('slidechanged', (event) => {
    if (event.indexh === 0) {
      Reveal.configure({ slideNumber: null });
      document.querySelector("div.has-logo > img.slide-logo").style.display = "none";
    }
    if (event.indexh === 1) {
      Reveal.configure({ slideNumber: 'c' });
      document.querySelector("div.has-logo > img.slide-logo").style.display = null;
    }
  });
</script>




</body></html>